{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "17e831fe",
   "metadata": {},
   "source": [
    "# Analyzing PBMC scATAC-seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1bce476",
   "metadata": {},
   "source": [
    "In this notebook, we will analyze PBMC scATAC-seq data. \n",
    "\n",
    "Following [vignette](https://stuartlab.org/signac/articles/pbmc_vignette) and a [video](https://www.youtube.com/watch?v=yEKZJVjc5DY&t=1033s) by bioinfomagician."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0d8f789",
   "metadata": {},
   "source": [
    "## Preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60949265",
   "metadata": {},
   "source": [
    "Here are the data we will use. Data can be obtained by either using `wget` command or from the [vignette](https://stuartlab.org/signac/articles/pbmc_vignette).\n",
    "\n",
    "``` \n",
    "wget https://cf.10xgenomics.com/samples/cell-atac/1.0.1/atac_v1_pbmc_10k/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5\n",
    "wget https://cf.10xgenomics.com/samples/cell-atac/1.0.1/atac_v1_pbmc_10k/atac_v1_pbmc_10k_singlecell.csv\n",
    "wget https://cf.10xgenomics.com/samples/cell-atac/1.0.1/atac_v1_pbmc_10k/atac_v1_pbmc_10k_fragments.tsv.gz\n",
    "wget https://cf.10xgenomics.com/samples/cell-atac/1.0.1/atac_v1_pbmc_10k/atac_v1_pbmc_10k_fragments.tsv.gz.tbi\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4685db4",
   "metadata": {},
   "source": [
    "We can save all the files in `'../data/atac'` after downloading."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "697194a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5'</li><li>'atac_v1_pbmc_10k_fragments.tsv.gz'</li><li>'atac_v1_pbmc_10k_fragments.tsv.gz.tbi'</li><li>'atac_v1_pbmc_10k_singlecell.csv'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'atac\\_v1\\_pbmc\\_10k\\_filtered\\_peak\\_bc\\_matrix.h5'\n",
       "\\item 'atac\\_v1\\_pbmc\\_10k\\_fragments.tsv.gz'\n",
       "\\item 'atac\\_v1\\_pbmc\\_10k\\_fragments.tsv.gz.tbi'\n",
       "\\item 'atac\\_v1\\_pbmc\\_10k\\_singlecell.csv'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5'\n",
       "2. 'atac_v1_pbmc_10k_fragments.tsv.gz'\n",
       "3. 'atac_v1_pbmc_10k_fragments.tsv.gz.tbi'\n",
       "4. 'atac_v1_pbmc_10k_singlecell.csv'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5\"\n",
       "[2] \"atac_v1_pbmc_10k_fragments.tsv.gz\"          \n",
       "[3] \"atac_v1_pbmc_10k_fragments.tsv.gz.tbi\"      \n",
       "[4] \"atac_v1_pbmc_10k_singlecell.csv\"            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "list.files(path = '../data/atac')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4106f14",
   "metadata": {},
   "source": [
    "Install packages if needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0b6f3fc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remotes::install_github(\"stuart-lab/signac\", ref=\"develop\")\n",
    "# install.packages(\"Matrix\", type = \"source\")\n",
    "# install.packages(\"irlba\", type = \"source\")\n",
    "# BiocManager::install(\"EnsDb.Hsapiens.v75\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4d2604b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, t\n",
      "\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.5.1     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32m%within%()\u001b[39m masks \u001b[34mIRanges\u001b[39m::%within%()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcollapse()\u001b[39m     masks \u001b[34mIRanges\u001b[39m::collapse()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcombine()\u001b[39m      masks \u001b[34mBiobase\u001b[39m::combine(), \u001b[34mBiocGenerics\u001b[39m::combine()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mdesc()\u001b[39m         masks \u001b[34mIRanges\u001b[39m::desc()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m       masks \u001b[34mensembldb\u001b[39m::filter(), \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m        masks \u001b[34mS4Vectors\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m          masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggplot2\u001b[39m::\u001b[32mPosition()\u001b[39m   masks \u001b[34mBiocGenerics\u001b[39m::Position(), \u001b[34mbase\u001b[39m::Position()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mreduce()\u001b[39m       masks \u001b[34mGenomicRanges\u001b[39m::reduce(), \u001b[34mIRanges\u001b[39m::reduce(), \u001b[34mSignac\u001b[39m::reduce()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond()\u001b[39m   masks \u001b[34mS4Vectors\u001b[39m::second()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond<-()\u001b[39m masks \u001b[34mS4Vectors\u001b[39m::second<-()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mselect()\u001b[39m       masks \u001b[34mensembldb\u001b[39m::select(), \u001b[34mAnnotationDbi\u001b[39m::select()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mslice()\u001b[39m        masks \u001b[34mIRanges\u001b[39m::slice()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n"
     ]
    }
   ],
   "source": [
    "library(Signac)\n",
    "library(Seurat)\n",
    "library(EnsDb.Hsapiens.v75)\n",
    "library(tidyverse)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73a41bb3",
   "metadata": {},
   "source": [
    "## Look at the data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebd43011",
   "metadata": {},
   "source": [
    "This is what frag file looks like."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4a5684f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>V1</th><th scope=col>V2</th><th scope=col>V3</th><th scope=col>V4</th><th scope=col>V5</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1</td><td>10066</td><td>10279</td><td>TTAGCTTAGGAGAACA-1</td><td>2</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1</td><td>10072</td><td>10279</td><td>TTAGCTTAGGAGAACA-1</td><td>2</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1</td><td>10079</td><td>10316</td><td>ATATTCCTCTTGTACT-1</td><td>2</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1</td><td>10084</td><td>10340</td><td>CGTACAAGTTACCCAA-1</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1</td><td>10085</td><td>10271</td><td>TGTGACAGTACAACGG-1</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1</td><td>10085</td><td>10339</td><td>CATGCCTTCTCTGACC-1</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & V1 & V2 & V3 & V4 & V5\\\\\n",
       "  & <chr> & <int> & <int> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & chr1 & 10066 & 10279 & TTAGCTTAGGAGAACA-1 & 2\\\\\n",
       "\t2 & chr1 & 10072 & 10279 & TTAGCTTAGGAGAACA-1 & 2\\\\\n",
       "\t3 & chr1 & 10079 & 10316 & ATATTCCTCTTGTACT-1 & 2\\\\\n",
       "\t4 & chr1 & 10084 & 10340 & CGTACAAGTTACCCAA-1 & 1\\\\\n",
       "\t5 & chr1 & 10085 & 10271 & TGTGACAGTACAACGG-1 & 1\\\\\n",
       "\t6 & chr1 & 10085 & 10339 & CATGCCTTCTCTGACC-1 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | V1 &lt;chr&gt; | V2 &lt;int&gt; | V3 &lt;int&gt; | V4 &lt;chr&gt; | V5 &lt;int&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | chr1 | 10066 | 10279 | TTAGCTTAGGAGAACA-1 | 2 |\n",
       "| 2 | chr1 | 10072 | 10279 | TTAGCTTAGGAGAACA-1 | 2 |\n",
       "| 3 | chr1 | 10079 | 10316 | ATATTCCTCTTGTACT-1 | 2 |\n",
       "| 4 | chr1 | 10084 | 10340 | CGTACAAGTTACCCAA-1 | 1 |\n",
       "| 5 | chr1 | 10085 | 10271 | TGTGACAGTACAACGG-1 | 1 |\n",
       "| 6 | chr1 | 10085 | 10339 | CATGCCTTCTCTGACC-1 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1   V2    V3    V4                 V5\n",
       "1 chr1 10066 10279 TTAGCTTAGGAGAACA-1 2 \n",
       "2 chr1 10072 10279 TTAGCTTAGGAGAACA-1 2 \n",
       "3 chr1 10079 10316 ATATTCCTCTTGTACT-1 2 \n",
       "4 chr1 10084 10340 CGTACAAGTTACCCAA-1 1 \n",
       "5 chr1 10085 10271 TGTGACAGTACAACGG-1 1 \n",
       "6 chr1 10085 10339 CATGCCTTCTCTGACC-1 1 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "frag.file <- read.delim('../data/atac/atac_v1_pbmc_10k_fragments.tsv.gz', header = F, nrows = 10)\n",
    "head(frag.file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27de8e7e",
   "metadata": {},
   "source": [
    "There are five columns in this data. Here is a quick overview on each column:\n",
    "\n",
    "    - First column: Reference genome used to identify chromosome.\n",
    "    - Second: Starting position of the fragment.\n",
    "    - Third: End position of the fragment.\n",
    "    - Fourth: Barcode fragment, which is used to identify sample. Same as `CB` tag in BAM file.\n",
    "    - Fifth: Total number of reads associated with this fragment.\n",
    "    \n",
    "More information about fragment data can be found on https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/output/fragments."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b2142c3",
   "metadata": {},
   "source": [
    "Let's take a look at this now. If you get an error saying `Error in Read10X_h5(\"../data/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5\"): Please install hdf5r to read HDF5 files`, install `libhdf5-dev` using \n",
    "\n",
    "`sudo apt-get install libhdf5-dev` (using terminal on debian/ubuntu)\n",
    "\n",
    "and \n",
    "\n",
    "`install.packages('hdf5r')` on R.\n",
    "\n",
    "You can also install it using github and devtool:\n",
    "``` R\n",
    "install.packages('devtools')\n",
    "devtools::install_github(\"hhoeflin/hdf5r\")\n",
    "```\n",
    "\n",
    "If you are using anaconda, you have to install `hdfr5` on conda using `mamba install hdf5` then install `hdf5r` on R."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4145a518",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Skipping install of 'hdf5r' from a github remote, the SHA1 (dc4774c0) has not changed since last install.\n",
      "  Use `force = TRUE` to force installation\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages('devtools')\n",
    "devtools::install_github(\"hhoeflin/hdf5r\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "173dcc72",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  [[ suppressing 10 column names 'AAACGAAAGAGCGAAA-1', 'AAACGAAAGAGTTTGA-1', 'AAACGAAAGCGAGCTA-1' ... ]]\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "10 x 10 sparse Matrix of class \"dgCMatrix\"\n",
       "                                      \n",
       "chr1:565107-565550 . . . . . . . . . .\n",
       "chr1:569174-569639 . . . . . . . . . .\n",
       "chr1:713460-714823 . . 2 8 . . 2 2 . .\n",
       "chr1:752422-753038 . . . . . . . . . .\n",
       "chr1:762106-763359 . . 4 2 . . . . 2 .\n",
       "chr1:779589-780271 . . . . . . . . . .\n",
       "chr1:793516-793741 . . . 1 . . . . . .\n",
       "chr1:801120-801338 . . . . . . . . . .\n",
       "chr1:804872-805761 . . . 4 . . . . . .\n",
       "chr1:839520-841123 . . 2 2 . . . . . ."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "counts <- Read10X_h5('../data/atac/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5')\n",
    "counts[1:10,1:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "328179ee",
   "metadata": {},
   "source": [
    "The first number is the range of open chromatin. Dots mean zeros. Numbers after the range represent `tn5` integration site in each cell that maps to the region. `tn5` is an enzyme that binds to open chromatin region. Learn more about `tn5`: https://www.youtube.com/watch?v=uuxpyhGNDsk. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01806277",
   "metadata": {},
   "source": [
    "`chrom_assay` has many information in it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "54fb646e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'ChromatinAssay' [package \"Signac\"] with 16 slots\n",
      "  ..@ ranges            :Formal class 'GRanges' [package \"GenomicRanges\"] with 7 slots\n",
      "  .. .. ..@ seqnames       :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. .. .. ..@ values         : Factor w/ 24 levels \"chr1\",\"chr2\",..: 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..@ lengths        : int [1:24] 8555 6858 5499 3515 4389 5269 4263 3677 3540 4137 ...\n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. ..@ ranges         :Formal class 'IRanges' [package \"IRanges\"] with 6 slots\n",
      "  .. .. .. .. ..@ start          : int [1:87561] 565107 569174 713460 752422 762106 779589 793516 801120 804872 839520 ...\n",
      "  .. .. .. .. ..@ width          : int [1:87561] 444 466 1364 617 1254 683 226 219 890 1604 ...\n",
      "  .. .. .. .. ..@ NAMES          : NULL\n",
      "  .. .. .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. ..@ strand         :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. .. .. ..@ values         : Factor w/ 3 levels \"+\",\"-\",\"*\": 3\n",
      "  .. .. .. .. ..@ lengths        : int 87561\n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. ..@ seqinfo        :Formal class 'Seqinfo' [package \"GenomeInfoDb\"] with 4 slots\n",
      "  .. .. .. .. ..@ seqnames   : chr [1:24] \"chr1\" \"chr2\" \"chr3\" \"chr4\" ...\n",
      "  .. .. .. .. ..@ seqlengths : int [1:24] NA NA NA NA NA NA NA NA NA NA ...\n",
      "  .. .. .. .. ..@ is_circular: logi [1:24] NA NA NA NA NA NA ...\n",
      "  .. .. .. .. ..@ genome     : chr [1:24] NA NA NA NA ...\n",
      "  .. .. ..@ elementMetadata:Formal class 'DFrame' [package \"S4Vectors\"] with 6 slots\n",
      "  .. .. .. .. ..@ rownames       : NULL\n",
      "  .. .. .. .. ..@ nrows          : int 87561\n",
      "  .. .. .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. .. .. ..@ listData       : Named list()\n",
      "  .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ motifs            : NULL\n",
      "  ..@ fragments         :List of 1\n",
      "  .. ..$ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. .. .. ..@ path : chr \"/home/kappa/git/goldfish/data/atac/atac_v1_pbmc_10k_fragments.tsv.gz\"\n",
      "  .. .. .. ..@ hash : chr [1:2] \"8e03bd72e952a7da7c1cce52f3696e6e\" \"146e556e1311573593bd6cfa6efefcc1\"\n",
      "  .. .. .. ..@ cells: Named chr [1:8728] \"AAACGAAAGAGCGAAA-1\" \"AAACGAAAGAGTTTGA-1\" \"AAACGAAAGCGAGCTA-1\" \"AAACGAAAGGCTTCGC-1\" ...\n",
      "  .. .. .. .. ..- attr(*, \"names\")= chr [1:8728] \"AAACGAAAGAGCGAAA-1\" \"AAACGAAAGAGTTTGA-1\" \"AAACGAAAGCGAGCTA-1\" \"AAACGAAAGGCTTCGC-1\" ...\n",
      "  ..@ seqinfo           : NULL\n",
      "  ..@ annotation        : NULL\n",
      "  ..@ bias              : NULL\n",
      "  ..@ positionEnrichment: list()\n",
      "  ..@ links             :Formal class 'GRanges' [package \"GenomicRanges\"] with 7 slots\n",
      "  .. .. ..@ seqnames       :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. .. .. ..@ values         : Factor w/ 0 levels: \n",
      "  .. .. .. .. ..@ lengths        : int(0) \n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. ..@ ranges         :Formal class 'IRanges' [package \"IRanges\"] with 6 slots\n",
      "  .. .. .. .. ..@ start          : int(0) \n",
      "  .. .. .. .. ..@ width          : int(0) \n",
      "  .. .. .. .. ..@ NAMES          : NULL\n",
      "  .. .. .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. ..@ strand         :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. .. .. ..@ values         : Factor w/ 3 levels \"+\",\"-\",\"*\": \n",
      "  .. .. .. .. ..@ lengths        : int(0) \n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. ..@ seqinfo        :Formal class 'Seqinfo' [package \"GenomeInfoDb\"] with 4 slots\n",
      "  .. .. .. .. ..@ seqnames   : chr(0) \n",
      "  .. .. .. .. ..@ seqlengths : int(0) \n",
      "  .. .. .. .. ..@ is_circular: logi(0) \n",
      "  .. .. .. .. ..@ genome     : chr(0) \n",
      "  .. .. ..@ elementMetadata:Formal class 'DFrame' [package \"S4Vectors\"] with 6 slots\n",
      "  .. .. .. .. ..@ rownames       : NULL\n",
      "  .. .. .. .. ..@ nrows          : int 0\n",
      "  .. .. .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. .. .. ..@ listData       : Named list()\n",
      "  .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ counts            :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. ..@ i       : int [1:52681465] 31 74 83 116 134 147 153 166 175 178 ...\n",
      "  .. .. ..@ p       : int [1:8729] 0 5546 12127 21340 51653 56574 61650 64340 73157 79801 ...\n",
      "  .. .. ..@ Dim     : int [1:2] 87561 8728\n",
      "  .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. ..$ : chr [1:87561] \"chr1-565107-565550\" \"chr1-569174-569639\" \"chr1-713460-714823\" \"chr1-752422-753038\" ...\n",
      "  .. .. .. ..$ : chr [1:8728] \"AAACGAAAGAGCGAAA-1\" \"AAACGAAAGAGTTTGA-1\" \"AAACGAAAGCGAGCTA-1\" \"AAACGAAAGGCTTCGC-1\" ...\n",
      "  .. .. ..@ x       : num [1:52681465] 2 2 2 1 2 2 2 6 2 2 ...\n",
      "  .. .. ..@ factors : list()\n",
      "  ..@ data              :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. ..@ i       : int [1:52681465] 31 74 83 116 134 147 153 166 175 178 ...\n",
      "  .. .. ..@ p       : int [1:8729] 0 5546 12127 21340 51653 56574 61650 64340 73157 79801 ...\n",
      "  .. .. ..@ Dim     : int [1:2] 87561 8728\n",
      "  .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. ..$ : chr [1:87561] \"chr1-565107-565550\" \"chr1-569174-569639\" \"chr1-713460-714823\" \"chr1-752422-753038\" ...\n",
      "  .. .. .. ..$ : chr [1:8728] \"AAACGAAAGAGCGAAA-1\" \"AAACGAAAGAGTTTGA-1\" \"AAACGAAAGCGAGCTA-1\" \"AAACGAAAGGCTTCGC-1\" ...\n",
      "  .. .. ..@ x       : num [1:52681465] 2 2 2 1 2 2 2 6 2 2 ...\n",
      "  .. .. ..@ factors : list()\n",
      "  ..@ scale.data        : num[0 , 0 ] \n",
      "  ..@ assay.orig        : NULL\n",
      "  ..@ var.features      : logi(0) \n",
      "  ..@ meta.features     :'data.frame':\t87561 obs. of  0 variables\n",
      "  ..@ misc              :List of 1\n",
      "  .. ..$ calcN: logi TRUE\n",
      "  ..@ key               : chr \"\"\n"
     ]
    }
   ],
   "source": [
    "chrom_assay <- CreateChromatinAssay(\n",
    "  counts = counts,\n",
    "  sep = c(\":\", \"-\"),\n",
    "  fragments = \"../data/atac/atac_v1_pbmc_10k_fragments.tsv.gz\",\n",
    "  min.cells = 10,\n",
    "  min.features = 200\n",
    ")\n",
    "\n",
    "str(chrom_assay)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d7ac4113",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 614496 × 16</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>total</th><th scope=col>duplicate</th><th scope=col>chimeric</th><th scope=col>unmapped</th><th scope=col>lowmapq</th><th scope=col>mitochondrial</th><th scope=col>passed_filters</th><th scope=col>cell_id</th><th scope=col>is__cell_barcode</th><th scope=col>TSS_fragments</th><th scope=col>DNase_sensitive_region_fragments</th><th scope=col>enhancer_region_fragments</th><th scope=col>promoter_region_fragments</th><th scope=col>on_target_fragments</th><th scope=col>blacklist_region_fragments</th><th scope=col>peak_region_fragments</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>NO_BARCODE</th><td>8335327</td><td>3922818</td><td>71559</td><td>746476</td><td>634014</td><td>22972</td><td>2937488</td><td>None</td><td>0</td><td>  0</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAAACGCC-1</th><td>      3</td><td>      1</td><td>    0</td><td>     1</td><td>     0</td><td>    0</td><td>      1</td><td>None</td><td>0</td><td>  1</td><td>   1</td><td>  1</td><td>  0</td><td>   1</td><td> 0</td><td>  1</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAAAGCAG-1</th><td>     14</td><td>      1</td><td>    0</td><td>     4</td><td>     2</td><td>    0</td><td>      7</td><td>None</td><td>0</td><td>  0</td><td>   2</td><td>  0</td><td>  0</td><td>   2</td><td> 0</td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAAAGGGT-1</th><td>      7</td><td>      1</td><td>    0</td><td>     1</td><td>     0</td><td>    0</td><td>      5</td><td>None</td><td>0</td><td>  2</td><td>   2</td><td>  0</td><td>  1</td><td>   2</td><td> 0</td><td>  1</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAAATACC-1</th><td>   9880</td><td>   5380</td><td>   79</td><td>   106</td><td>  1120</td><td>    6</td><td>   3189</td><td>None</td><td>0</td><td>386</td><td>1623</td><td>297</td><td>107</td><td>1758</td><td>16</td><td>262</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAAATCTG-1</th><td>      2</td><td>      0</td><td>    0</td><td>     0</td><td>     0</td><td>    0</td><td>      2</td><td>None</td><td>0</td><td>  1</td><td>   2</td><td>  1</td><td>  1</td><td>   2</td><td> 0</td><td>  2</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAAATGGG-1</th><td>      4</td><td>      2</td><td>    0</td><td>     1</td><td>     0</td><td>    0</td><td>      1</td><td>None</td><td>0</td><td>  0</td><td>   1</td><td>  1</td><td>  0</td><td>   1</td><td> 0</td><td>  1</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAAATTCG-1</th><td>      2</td><td>      0</td><td>    0</td><td>     0</td><td>     1</td><td>    0</td><td>      1</td><td>None</td><td>0</td><td>  0</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAACCATA-1</th><td>     12</td><td>      3</td><td>    0</td><td>     8</td><td>     0</td><td>    0</td><td>      1</td><td>None</td><td>0</td><td>  0</td><td>   1</td><td>  0</td><td>  0</td><td>   1</td><td> 0</td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAACCCGA-1</th><td>      3</td><td>      1</td><td>    0</td><td>     2</td><td>     0</td><td>    0</td><td>      0</td><td>None</td><td>0</td><td>  0</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAACGACC-1</th><td>      2</td><td>      0</td><td>    0</td><td>     0</td><td>     0</td><td>    0</td><td>      2</td><td>None</td><td>0</td><td>  2</td><td>   2</td><td>  1</td><td>  1</td><td>   2</td><td> 0</td><td>  1</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAACGCCA-1</th><td>      7</td><td>      0</td><td>    0</td><td>     1</td><td>     0</td><td>    0</td><td>      6</td><td>None</td><td>0</td><td>  0</td><td>   4</td><td>  0</td><td>  0</td><td>   4</td><td> 0</td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAACGTCG-1</th><td>      3</td><td>      0</td><td>    0</td><td>     1</td><td>     0</td><td>    0</td><td>      2</td><td>None</td><td>0</td><td>  0</td><td>   1</td><td>  1</td><td>  0</td><td>   1</td><td> 0</td><td>  1</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAACGTTA-1</th><td>     11</td><td>      1</td><td>    0</td><td>     2</td><td>     2</td><td>    0</td><td>      6</td><td>None</td><td>0</td><td>  2</td><td>   6</td><td>  1</td><td>  2</td><td>   6</td><td> 0</td><td>  4</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAACTAAC-1</th><td>      7</td><td>      0</td><td>    0</td><td>     5</td><td>     1</td><td>    0</td><td>      1</td><td>None</td><td>0</td><td>  1</td><td>   1</td><td>  0</td><td>  1</td><td>   1</td><td> 0</td><td>  1</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAACTCCT-1</th><td>      2</td><td>      2</td><td>    0</td><td>     0</td><td>     0</td><td>    0</td><td>      0</td><td>None</td><td>0</td><td>  0</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAAGACTT-1</th><td>      3</td><td>      2</td><td>    0</td><td>     0</td><td>     0</td><td>    0</td><td>      1</td><td>None</td><td>0</td><td>  0</td><td>   1</td><td>  1</td><td>  0</td><td>   1</td><td> 0</td><td>  1</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAAGAGTG-1</th><td>      4</td><td>      1</td><td>    0</td><td>     0</td><td>     0</td><td>    0</td><td>      3</td><td>None</td><td>0</td><td>  0</td><td>   1</td><td>  1</td><td>  0</td><td>   2</td><td> 0</td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAAGCCGT-1</th><td>      2</td><td>      0</td><td>    0</td><td>     0</td><td>     1</td><td>    0</td><td>      1</td><td>None</td><td>0</td><td>  0</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAAGCCTG-1</th><td>      7</td><td>      0</td><td>    0</td><td>     0</td><td>     1</td><td>    0</td><td>      6</td><td>None</td><td>0</td><td>  0</td><td>   4</td><td>  0</td><td>  0</td><td>   4</td><td> 0</td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAAGGGCG-1</th><td>     10</td><td>      0</td><td>    2</td><td>     1</td><td>     1</td><td>    0</td><td>      6</td><td>None</td><td>0</td><td>  0</td><td>   3</td><td>  1</td><td>  0</td><td>   3</td><td> 0</td><td>  2</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAATACTG-1</th><td>     10</td><td>      1</td><td>    0</td><td>     2</td><td>     2</td><td>    0</td><td>      5</td><td>None</td><td>0</td><td>  0</td><td>   4</td><td>  0</td><td>  0</td><td>   4</td><td> 0</td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAATAGTC-1</th><td>    464</td><td>     79</td><td>    0</td><td>    10</td><td>    40</td><td>    0</td><td>    335</td><td>None</td><td>0</td><td> 43</td><td> 176</td><td> 31</td><td>  4</td><td> 194</td><td> 2</td><td> 19</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAATCAAC-1</th><td>     11</td><td>      2</td><td>    0</td><td>     4</td><td>     1</td><td>    0</td><td>      4</td><td>None</td><td>0</td><td>  0</td><td>   0</td><td>  1</td><td>  0</td><td>   1</td><td> 0</td><td>  1</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAATCAGT-1</th><td>      4</td><td>      1</td><td>    0</td><td>     1</td><td>     0</td><td>    0</td><td>      2</td><td>None</td><td>0</td><td>  0</td><td>   1</td><td>  1</td><td>  0</td><td>   1</td><td> 0</td><td>  1</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGAATGCTA-1</th><td>      5</td><td>      2</td><td>    0</td><td>     0</td><td>     2</td><td>    0</td><td>      1</td><td>None</td><td>0</td><td>  0</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGACAACGC-1</th><td>     30</td><td>      6</td><td>    0</td><td>     6</td><td>     0</td><td>    0</td><td>     18</td><td>None</td><td>0</td><td>  8</td><td>  15</td><td>  2</td><td>  7</td><td>  16</td><td> 0</td><td> 11</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGACACAAT-1</th><td>      5</td><td>      0</td><td>    0</td><td>     2</td><td>     0</td><td>    0</td><td>      3</td><td>None</td><td>0</td><td>  1</td><td>   1</td><td>  0</td><td>  1</td><td>   1</td><td> 0</td><td>  1</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGACACGGT-1</th><td>   1643</td><td>    417</td><td>   11</td><td>    12</td><td>   214</td><td>    8</td><td>    981</td><td>None</td><td>0</td><td> 97</td><td> 502</td><td>100</td><td> 17</td><td> 540</td><td> 5</td><td> 82</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGACACTTC-1</th><td>     11</td><td>      3</td><td>    0</td><td>     2</td><td>     0</td><td>    0</td><td>      6</td><td>None</td><td>0</td><td>  1</td><td>   4</td><td>  2</td><td>  0</td><td>   4</td><td> 0</td><td>  1</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTGGGCGT-1</th><td>    4</td><td>   0</td><td>  0</td><td>  0</td><td>   1</td><td> 0</td><td>    3</td><td>None      </td><td>0</td><td>   0</td><td>    2</td><td>   1</td><td>   0</td><td>    2</td><td> 0</td><td>   1</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTGGTACA-1</th><td>   56</td><td>  14</td><td>  0</td><td>  1</td><td>   1</td><td> 0</td><td>   40</td><td>None      </td><td>0</td><td>  29</td><td>   34</td><td>   5</td><td>  25</td><td>   38</td><td> 0</td><td>  35</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTGTCGGG-1</th><td> 1723</td><td> 505</td><td> 15</td><td> 15</td><td> 190</td><td> 0</td><td>  998</td><td>None      </td><td>0</td><td> 117</td><td>  462</td><td>  89</td><td>  25</td><td>  514</td><td>12</td><td>  76</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTGTCGTC-1</th><td>   12</td><td>   9</td><td>  0</td><td>  0</td><td>   1</td><td> 0</td><td>    2</td><td>None      </td><td>0</td><td>   1</td><td>    2</td><td>   1</td><td>   1</td><td>    2</td><td> 0</td><td>   2</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTGTGTCC-1</th><td>    5</td><td>   1</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    4</td><td>None      </td><td>0</td><td>   1</td><td>    1</td><td>   0</td><td>   1</td><td>    1</td><td> 0</td><td>   1</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTGTGTGA-1</th><td> 2460</td><td> 678</td><td> 17</td><td> 24</td><td> 338</td><td> 2</td><td> 1401</td><td>None      </td><td>0</td><td> 134</td><td>  652</td><td>  99</td><td>  33</td><td>  703</td><td>11</td><td>  92</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTGTTTAC-1</th><td>    2</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    2</td><td>None      </td><td>0</td><td>   0</td><td>    1</td><td>   0</td><td>   0</td><td>    1</td><td> 0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTAAGCG-1</th><td>    1</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    1</td><td>None      </td><td>0</td><td>   1</td><td>    0</td><td>   0</td><td>   0</td><td>    1</td><td> 0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTAATCC-1</th><td>    2</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    2</td><td>None      </td><td>0</td><td>   1</td><td>    1</td><td>   0</td><td>   1</td><td>    1</td><td> 0</td><td>   1</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTACCTA-1</th><td>    1</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    1</td><td>None      </td><td>0</td><td>   0</td><td>    0</td><td>   0</td><td>   0</td><td>    0</td><td> 0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTACGGA-1</th><td>   10</td><td>   1</td><td>  0</td><td>  1</td><td>   1</td><td> 1</td><td>    6</td><td>None      </td><td>0</td><td>   2</td><td>    1</td><td>   1</td><td>   0</td><td>    3</td><td> 0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTACTCA-1</th><td>    1</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    1</td><td>None      </td><td>0</td><td>   0</td><td>    1</td><td>   1</td><td>   0</td><td>    1</td><td> 0</td><td>   1</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTAGTGG-1</th><td>    2</td><td>   0</td><td>  0</td><td>  0</td><td>   1</td><td> 0</td><td>    1</td><td>None      </td><td>0</td><td>   1</td><td>    1</td><td>   1</td><td>   0</td><td>    1</td><td> 0</td><td>   1</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTATCAC-1</th><td>    3</td><td>   2</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    1</td><td>None      </td><td>0</td><td>   0</td><td>    0</td><td>   0</td><td>   0</td><td>    0</td><td> 0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTCACTA-1</th><td>    1</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    1</td><td>None      </td><td>0</td><td>   0</td><td>    1</td><td>   0</td><td>   0</td><td>    1</td><td> 0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTCAGAG-1</th><td>    1</td><td>   0</td><td>  0</td><td>  0</td><td>   1</td><td> 0</td><td>    0</td><td>None      </td><td>0</td><td>   0</td><td>    0</td><td>   0</td><td>   0</td><td>    0</td><td> 0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTCATAC-1</th><td>    4</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    4</td><td>None      </td><td>0</td><td>   0</td><td>    2</td><td>   1</td><td>   0</td><td>    2</td><td> 0</td><td>   1</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTCCACG-1</th><td>    4</td><td>   1</td><td>  0</td><td>  0</td><td>   1</td><td> 0</td><td>    2</td><td>None      </td><td>0</td><td>   2</td><td>    2</td><td>   0</td><td>   2</td><td>    2</td><td> 0</td><td>   2</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTCCAGC-1</th><td>    1</td><td>   1</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    0</td><td>None      </td><td>0</td><td>   0</td><td>    0</td><td>   0</td><td>   0</td><td>    0</td><td> 0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTCCGTG-1</th><td>   19</td><td>   8</td><td>  0</td><td>  1</td><td>   2</td><td> 0</td><td>    8</td><td>None      </td><td>0</td><td>   4</td><td>    5</td><td>   0</td><td>   2</td><td>    6</td><td> 0</td><td>   3</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTCGTTA-1</th><td>   45</td><td>  11</td><td>  0</td><td>  1</td><td>   4</td><td> 0</td><td>   29</td><td>None      </td><td>0</td><td>   9</td><td>   20</td><td>   4</td><td>   8</td><td>   20</td><td> 0</td><td>  16</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTCTCTC-1</th><td>    1</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    1</td><td>None      </td><td>0</td><td>   0</td><td>    0</td><td>   0</td><td>   0</td><td>    0</td><td> 0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTCTTCC-1</th><td>   14</td><td>   7</td><td>  0</td><td>  1</td><td>   3</td><td> 0</td><td>    3</td><td>None      </td><td>0</td><td>   2</td><td>    3</td><td>   0</td><td>   2</td><td>    3</td><td> 0</td><td>   2</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTGCGCT-1</th><td>   32</td><td>   2</td><td>  1</td><td>  1</td><td>   3</td><td> 1</td><td>   24</td><td>None      </td><td>0</td><td>  12</td><td>   17</td><td>   2</td><td>   9</td><td>   18</td><td> 0</td><td>  12</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTGGCCT-1</th><td>    3</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    3</td><td>None      </td><td>0</td><td>   0</td><td>    3</td><td>   1</td><td>   0</td><td>    3</td><td> 0</td><td>   1</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTGTACT-1</th><td>   47</td><td>   2</td><td>  0</td><td>  0</td><td>   4</td><td> 0</td><td>   41</td><td>None      </td><td>0</td><td>  11</td><td>   31</td><td>   9</td><td>   8</td><td>   32</td><td> 0</td><td>  20</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTGTCGC-1</th><td>    1</td><td>   0</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    1</td><td>None      </td><td>0</td><td>   0</td><td>    0</td><td>   0</td><td>   0</td><td>    0</td><td> 0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTGTGCC-1</th><td>23379</td><td>9625</td><td>160</td><td>115</td><td>1232</td><td>20</td><td>12227</td><td>_cell_8727</td><td>1</td><td>6321</td><td>10005</td><td>2560</td><td>5242</td><td>10523</td><td>24</td><td>8309</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTTATCG-1</th><td>    2</td><td>   1</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    1</td><td>None      </td><td>0</td><td>   0</td><td>    0</td><td>   0</td><td>   0</td><td>    0</td><td> 0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTTGCAT-1</th><td>    3</td><td>   1</td><td>  0</td><td>  0</td><td>   0</td><td> 0</td><td>    2</td><td>None      </td><td>0</td><td>   1</td><td>    2</td><td>   0</td><td>   1</td><td>    2</td><td> 0</td><td>   1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 614496 × 16\n",
       "\\begin{tabular}{r|llllllllllllllll}\n",
       "  & total & duplicate & chimeric & unmapped & lowmapq & mitochondrial & passed\\_filters & cell\\_id & is\\_\\_cell\\_barcode & TSS\\_fragments & DNase\\_sensitive\\_region\\_fragments & enhancer\\_region\\_fragments & promoter\\_region\\_fragments & on\\_target\\_fragments & blacklist\\_region\\_fragments & peak\\_region\\_fragments\\\\\n",
       "  & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <chr> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\tNO\\_BARCODE & 8335327 & 3922818 & 71559 & 746476 & 634014 & 22972 & 2937488 & None & 0 &   0 &    0 &   0 &   0 &    0 &  0 &   0\\\\\n",
       "\tAAACGAAAGAAACGCC-1 &       3 &       1 &     0 &      1 &      0 &     0 &       1 & None & 0 &   1 &    1 &   1 &   0 &    1 &  0 &   1\\\\\n",
       "\tAAACGAAAGAAAGCAG-1 &      14 &       1 &     0 &      4 &      2 &     0 &       7 & None & 0 &   0 &    2 &   0 &   0 &    2 &  0 &   0\\\\\n",
       "\tAAACGAAAGAAAGGGT-1 &       7 &       1 &     0 &      1 &      0 &     0 &       5 & None & 0 &   2 &    2 &   0 &   1 &    2 &  0 &   1\\\\\n",
       "\tAAACGAAAGAAATACC-1 &    9880 &    5380 &    79 &    106 &   1120 &     6 &    3189 & None & 0 & 386 & 1623 & 297 & 107 & 1758 & 16 & 262\\\\\n",
       "\tAAACGAAAGAAATCTG-1 &       2 &       0 &     0 &      0 &      0 &     0 &       2 & None & 0 &   1 &    2 &   1 &   1 &    2 &  0 &   2\\\\\n",
       "\tAAACGAAAGAAATGGG-1 &       4 &       2 &     0 &      1 &      0 &     0 &       1 & None & 0 &   0 &    1 &   1 &   0 &    1 &  0 &   1\\\\\n",
       "\tAAACGAAAGAAATTCG-1 &       2 &       0 &     0 &      0 &      1 &     0 &       1 & None & 0 &   0 &    0 &   0 &   0 &    0 &  0 &   0\\\\\n",
       "\tAAACGAAAGAACCATA-1 &      12 &       3 &     0 &      8 &      0 &     0 &       1 & None & 0 &   0 &    1 &   0 &   0 &    1 &  0 &   0\\\\\n",
       "\tAAACGAAAGAACCCGA-1 &       3 &       1 &     0 &      2 &      0 &     0 &       0 & None & 0 &   0 &    0 &   0 &   0 &    0 &  0 &   0\\\\\n",
       "\tAAACGAAAGAACGACC-1 &       2 &       0 &     0 &      0 &      0 &     0 &       2 & None & 0 &   2 &    2 &   1 &   1 &    2 &  0 &   1\\\\\n",
       "\tAAACGAAAGAACGCCA-1 &       7 &       0 &     0 &      1 &      0 &     0 &       6 & None & 0 &   0 &    4 &   0 &   0 &    4 &  0 &   0\\\\\n",
       "\tAAACGAAAGAACGTCG-1 &       3 &       0 &     0 &      1 &      0 &     0 &       2 & None & 0 &   0 &    1 &   1 &   0 &    1 &  0 &   1\\\\\n",
       "\tAAACGAAAGAACGTTA-1 &      11 &       1 &     0 &      2 &      2 &     0 &       6 & None & 0 &   2 &    6 &   1 &   2 &    6 &  0 &   4\\\\\n",
       "\tAAACGAAAGAACTAAC-1 &       7 &       0 &     0 &      5 &      1 &     0 &       1 & None & 0 &   1 &    1 &   0 &   1 &    1 &  0 &   1\\\\\n",
       "\tAAACGAAAGAACTCCT-1 &       2 &       2 &     0 &      0 &      0 &     0 &       0 & None & 0 &   0 &    0 &   0 &   0 &    0 &  0 &   0\\\\\n",
       "\tAAACGAAAGAAGACTT-1 &       3 &       2 &     0 &      0 &      0 &     0 &       1 & None & 0 &   0 &    1 &   1 &   0 &    1 &  0 &   1\\\\\n",
       "\tAAACGAAAGAAGAGTG-1 &       4 &       1 &     0 &      0 &      0 &     0 &       3 & None & 0 &   0 &    1 &   1 &   0 &    2 &  0 &   0\\\\\n",
       "\tAAACGAAAGAAGCCGT-1 &       2 &       0 &     0 &      0 &      1 &     0 &       1 & None & 0 &   0 &    0 &   0 &   0 &    0 &  0 &   0\\\\\n",
       "\tAAACGAAAGAAGCCTG-1 &       7 &       0 &     0 &      0 &      1 &     0 &       6 & None & 0 &   0 &    4 &   0 &   0 &    4 &  0 &   0\\\\\n",
       "\tAAACGAAAGAAGGGCG-1 &      10 &       0 &     2 &      1 &      1 &     0 &       6 & None & 0 &   0 &    3 &   1 &   0 &    3 &  0 &   2\\\\\n",
       "\tAAACGAAAGAATACTG-1 &      10 &       1 &     0 &      2 &      2 &     0 &       5 & None & 0 &   0 &    4 &   0 &   0 &    4 &  0 &   0\\\\\n",
       "\tAAACGAAAGAATAGTC-1 &     464 &      79 &     0 &     10 &     40 &     0 &     335 & None & 0 &  43 &  176 &  31 &   4 &  194 &  2 &  19\\\\\n",
       "\tAAACGAAAGAATCAAC-1 &      11 &       2 &     0 &      4 &      1 &     0 &       4 & None & 0 &   0 &    0 &   1 &   0 &    1 &  0 &   1\\\\\n",
       "\tAAACGAAAGAATCAGT-1 &       4 &       1 &     0 &      1 &      0 &     0 &       2 & None & 0 &   0 &    1 &   1 &   0 &    1 &  0 &   1\\\\\n",
       "\tAAACGAAAGAATGCTA-1 &       5 &       2 &     0 &      0 &      2 &     0 &       1 & None & 0 &   0 &    0 &   0 &   0 &    0 &  0 &   0\\\\\n",
       "\tAAACGAAAGACAACGC-1 &      30 &       6 &     0 &      6 &      0 &     0 &      18 & None & 0 &   8 &   15 &   2 &   7 &   16 &  0 &  11\\\\\n",
       "\tAAACGAAAGACACAAT-1 &       5 &       0 &     0 &      2 &      0 &     0 &       3 & None & 0 &   1 &    1 &   0 &   1 &    1 &  0 &   1\\\\\n",
       "\tAAACGAAAGACACGGT-1 &    1643 &     417 &    11 &     12 &    214 &     8 &     981 & None & 0 &  97 &  502 & 100 &  17 &  540 &  5 &  82\\\\\n",
       "\tAAACGAAAGACACTTC-1 &      11 &       3 &     0 &      2 &      0 &     0 &       6 & None & 0 &   1 &    4 &   2 &   0 &    4 &  0 &   1\\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\tTTTGTGTTCTGGGCGT-1 &     4 &    0 &   0 &   0 &    1 &  0 &     3 & None       & 0 &    0 &     2 &    1 &    0 &     2 &  0 &    1\\\\\n",
       "\tTTTGTGTTCTGGTACA-1 &    56 &   14 &   0 &   1 &    1 &  0 &    40 & None       & 0 &   29 &    34 &    5 &   25 &    38 &  0 &   35\\\\\n",
       "\tTTTGTGTTCTGTCGGG-1 &  1723 &  505 &  15 &  15 &  190 &  0 &   998 & None       & 0 &  117 &   462 &   89 &   25 &   514 & 12 &   76\\\\\n",
       "\tTTTGTGTTCTGTCGTC-1 &    12 &    9 &   0 &   0 &    1 &  0 &     2 & None       & 0 &    1 &     2 &    1 &    1 &     2 &  0 &    2\\\\\n",
       "\tTTTGTGTTCTGTGTCC-1 &     5 &    1 &   0 &   0 &    0 &  0 &     4 & None       & 0 &    1 &     1 &    0 &    1 &     1 &  0 &    1\\\\\n",
       "\tTTTGTGTTCTGTGTGA-1 &  2460 &  678 &  17 &  24 &  338 &  2 &  1401 & None       & 0 &  134 &   652 &   99 &   33 &   703 & 11 &   92\\\\\n",
       "\tTTTGTGTTCTGTTTAC-1 &     2 &    0 &   0 &   0 &    0 &  0 &     2 & None       & 0 &    0 &     1 &    0 &    0 &     1 &  0 &    0\\\\\n",
       "\tTTTGTGTTCTTAAGCG-1 &     1 &    0 &   0 &   0 &    0 &  0 &     1 & None       & 0 &    1 &     0 &    0 &    0 &     1 &  0 &    0\\\\\n",
       "\tTTTGTGTTCTTAATCC-1 &     2 &    0 &   0 &   0 &    0 &  0 &     2 & None       & 0 &    1 &     1 &    0 &    1 &     1 &  0 &    1\\\\\n",
       "\tTTTGTGTTCTTACCTA-1 &     1 &    0 &   0 &   0 &    0 &  0 &     1 & None       & 0 &    0 &     0 &    0 &    0 &     0 &  0 &    0\\\\\n",
       "\tTTTGTGTTCTTACGGA-1 &    10 &    1 &   0 &   1 &    1 &  1 &     6 & None       & 0 &    2 &     1 &    1 &    0 &     3 &  0 &    0\\\\\n",
       "\tTTTGTGTTCTTACTCA-1 &     1 &    0 &   0 &   0 &    0 &  0 &     1 & None       & 0 &    0 &     1 &    1 &    0 &     1 &  0 &    1\\\\\n",
       "\tTTTGTGTTCTTAGTGG-1 &     2 &    0 &   0 &   0 &    1 &  0 &     1 & None       & 0 &    1 &     1 &    1 &    0 &     1 &  0 &    1\\\\\n",
       "\tTTTGTGTTCTTATCAC-1 &     3 &    2 &   0 &   0 &    0 &  0 &     1 & None       & 0 &    0 &     0 &    0 &    0 &     0 &  0 &    0\\\\\n",
       "\tTTTGTGTTCTTCACTA-1 &     1 &    0 &   0 &   0 &    0 &  0 &     1 & None       & 0 &    0 &     1 &    0 &    0 &     1 &  0 &    0\\\\\n",
       "\tTTTGTGTTCTTCAGAG-1 &     1 &    0 &   0 &   0 &    1 &  0 &     0 & None       & 0 &    0 &     0 &    0 &    0 &     0 &  0 &    0\\\\\n",
       "\tTTTGTGTTCTTCATAC-1 &     4 &    0 &   0 &   0 &    0 &  0 &     4 & None       & 0 &    0 &     2 &    1 &    0 &     2 &  0 &    1\\\\\n",
       "\tTTTGTGTTCTTCCACG-1 &     4 &    1 &   0 &   0 &    1 &  0 &     2 & None       & 0 &    2 &     2 &    0 &    2 &     2 &  0 &    2\\\\\n",
       "\tTTTGTGTTCTTCCAGC-1 &     1 &    1 &   0 &   0 &    0 &  0 &     0 & None       & 0 &    0 &     0 &    0 &    0 &     0 &  0 &    0\\\\\n",
       "\tTTTGTGTTCTTCCGTG-1 &    19 &    8 &   0 &   1 &    2 &  0 &     8 & None       & 0 &    4 &     5 &    0 &    2 &     6 &  0 &    3\\\\\n",
       "\tTTTGTGTTCTTCGTTA-1 &    45 &   11 &   0 &   1 &    4 &  0 &    29 & None       & 0 &    9 &    20 &    4 &    8 &    20 &  0 &   16\\\\\n",
       "\tTTTGTGTTCTTCTCTC-1 &     1 &    0 &   0 &   0 &    0 &  0 &     1 & None       & 0 &    0 &     0 &    0 &    0 &     0 &  0 &    0\\\\\n",
       "\tTTTGTGTTCTTCTTCC-1 &    14 &    7 &   0 &   1 &    3 &  0 &     3 & None       & 0 &    2 &     3 &    0 &    2 &     3 &  0 &    2\\\\\n",
       "\tTTTGTGTTCTTGCGCT-1 &    32 &    2 &   1 &   1 &    3 &  1 &    24 & None       & 0 &   12 &    17 &    2 &    9 &    18 &  0 &   12\\\\\n",
       "\tTTTGTGTTCTTGGCCT-1 &     3 &    0 &   0 &   0 &    0 &  0 &     3 & None       & 0 &    0 &     3 &    1 &    0 &     3 &  0 &    1\\\\\n",
       "\tTTTGTGTTCTTGTACT-1 &    47 &    2 &   0 &   0 &    4 &  0 &    41 & None       & 0 &   11 &    31 &    9 &    8 &    32 &  0 &   20\\\\\n",
       "\tTTTGTGTTCTTGTCGC-1 &     1 &    0 &   0 &   0 &    0 &  0 &     1 & None       & 0 &    0 &     0 &    0 &    0 &     0 &  0 &    0\\\\\n",
       "\tTTTGTGTTCTTGTGCC-1 & 23379 & 9625 & 160 & 115 & 1232 & 20 & 12227 & \\_cell\\_8727 & 1 & 6321 & 10005 & 2560 & 5242 & 10523 & 24 & 8309\\\\\n",
       "\tTTTGTGTTCTTTATCG-1 &     2 &    1 &   0 &   0 &    0 &  0 &     1 & None       & 0 &    0 &     0 &    0 &    0 &     0 &  0 &    0\\\\\n",
       "\tTTTGTGTTCTTTGCAT-1 &     3 &    1 &   0 &   0 &    0 &  0 &     2 & None       & 0 &    1 &     2 &    0 &    1 &     2 &  0 &    1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 614496 × 16\n",
       "\n",
       "| <!--/--> | total &lt;int&gt; | duplicate &lt;int&gt; | chimeric &lt;int&gt; | unmapped &lt;int&gt; | lowmapq &lt;int&gt; | mitochondrial &lt;int&gt; | passed_filters &lt;int&gt; | cell_id &lt;chr&gt; | is__cell_barcode &lt;int&gt; | TSS_fragments &lt;int&gt; | DNase_sensitive_region_fragments &lt;int&gt; | enhancer_region_fragments &lt;int&gt; | promoter_region_fragments &lt;int&gt; | on_target_fragments &lt;int&gt; | blacklist_region_fragments &lt;int&gt; | peak_region_fragments &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| NO_BARCODE | 8335327 | 3922818 | 71559 | 746476 | 634014 | 22972 | 2937488 | None | 0 |   0 |    0 |   0 |   0 |    0 |  0 |   0 |\n",
       "| AAACGAAAGAAACGCC-1 |       3 |       1 |     0 |      1 |      0 |     0 |       1 | None | 0 |   1 |    1 |   1 |   0 |    1 |  0 |   1 |\n",
       "| AAACGAAAGAAAGCAG-1 |      14 |       1 |     0 |      4 |      2 |     0 |       7 | None | 0 |   0 |    2 |   0 |   0 |    2 |  0 |   0 |\n",
       "| AAACGAAAGAAAGGGT-1 |       7 |       1 |     0 |      1 |      0 |     0 |       5 | None | 0 |   2 |    2 |   0 |   1 |    2 |  0 |   1 |\n",
       "| AAACGAAAGAAATACC-1 |    9880 |    5380 |    79 |    106 |   1120 |     6 |    3189 | None | 0 | 386 | 1623 | 297 | 107 | 1758 | 16 | 262 |\n",
       "| AAACGAAAGAAATCTG-1 |       2 |       0 |     0 |      0 |      0 |     0 |       2 | None | 0 |   1 |    2 |   1 |   1 |    2 |  0 |   2 |\n",
       "| AAACGAAAGAAATGGG-1 |       4 |       2 |     0 |      1 |      0 |     0 |       1 | None | 0 |   0 |    1 |   1 |   0 |    1 |  0 |   1 |\n",
       "| AAACGAAAGAAATTCG-1 |       2 |       0 |     0 |      0 |      1 |     0 |       1 | None | 0 |   0 |    0 |   0 |   0 |    0 |  0 |   0 |\n",
       "| AAACGAAAGAACCATA-1 |      12 |       3 |     0 |      8 |      0 |     0 |       1 | None | 0 |   0 |    1 |   0 |   0 |    1 |  0 |   0 |\n",
       "| AAACGAAAGAACCCGA-1 |       3 |       1 |     0 |      2 |      0 |     0 |       0 | None | 0 |   0 |    0 |   0 |   0 |    0 |  0 |   0 |\n",
       "| AAACGAAAGAACGACC-1 |       2 |       0 |     0 |      0 |      0 |     0 |       2 | None | 0 |   2 |    2 |   1 |   1 |    2 |  0 |   1 |\n",
       "| AAACGAAAGAACGCCA-1 |       7 |       0 |     0 |      1 |      0 |     0 |       6 | None | 0 |   0 |    4 |   0 |   0 |    4 |  0 |   0 |\n",
       "| AAACGAAAGAACGTCG-1 |       3 |       0 |     0 |      1 |      0 |     0 |       2 | None | 0 |   0 |    1 |   1 |   0 |    1 |  0 |   1 |\n",
       "| AAACGAAAGAACGTTA-1 |      11 |       1 |     0 |      2 |      2 |     0 |       6 | None | 0 |   2 |    6 |   1 |   2 |    6 |  0 |   4 |\n",
       "| AAACGAAAGAACTAAC-1 |       7 |       0 |     0 |      5 |      1 |     0 |       1 | None | 0 |   1 |    1 |   0 |   1 |    1 |  0 |   1 |\n",
       "| AAACGAAAGAACTCCT-1 |       2 |       2 |     0 |      0 |      0 |     0 |       0 | None | 0 |   0 |    0 |   0 |   0 |    0 |  0 |   0 |\n",
       "| AAACGAAAGAAGACTT-1 |       3 |       2 |     0 |      0 |      0 |     0 |       1 | None | 0 |   0 |    1 |   1 |   0 |    1 |  0 |   1 |\n",
       "| AAACGAAAGAAGAGTG-1 |       4 |       1 |     0 |      0 |      0 |     0 |       3 | None | 0 |   0 |    1 |   1 |   0 |    2 |  0 |   0 |\n",
       "| AAACGAAAGAAGCCGT-1 |       2 |       0 |     0 |      0 |      1 |     0 |       1 | None | 0 |   0 |    0 |   0 |   0 |    0 |  0 |   0 |\n",
       "| AAACGAAAGAAGCCTG-1 |       7 |       0 |     0 |      0 |      1 |     0 |       6 | None | 0 |   0 |    4 |   0 |   0 |    4 |  0 |   0 |\n",
       "| AAACGAAAGAAGGGCG-1 |      10 |       0 |     2 |      1 |      1 |     0 |       6 | None | 0 |   0 |    3 |   1 |   0 |    3 |  0 |   2 |\n",
       "| AAACGAAAGAATACTG-1 |      10 |       1 |     0 |      2 |      2 |     0 |       5 | None | 0 |   0 |    4 |   0 |   0 |    4 |  0 |   0 |\n",
       "| AAACGAAAGAATAGTC-1 |     464 |      79 |     0 |     10 |     40 |     0 |     335 | None | 0 |  43 |  176 |  31 |   4 |  194 |  2 |  19 |\n",
       "| AAACGAAAGAATCAAC-1 |      11 |       2 |     0 |      4 |      1 |     0 |       4 | None | 0 |   0 |    0 |   1 |   0 |    1 |  0 |   1 |\n",
       "| AAACGAAAGAATCAGT-1 |       4 |       1 |     0 |      1 |      0 |     0 |       2 | None | 0 |   0 |    1 |   1 |   0 |    1 |  0 |   1 |\n",
       "| AAACGAAAGAATGCTA-1 |       5 |       2 |     0 |      0 |      2 |     0 |       1 | None | 0 |   0 |    0 |   0 |   0 |    0 |  0 |   0 |\n",
       "| AAACGAAAGACAACGC-1 |      30 |       6 |     0 |      6 |      0 |     0 |      18 | None | 0 |   8 |   15 |   2 |   7 |   16 |  0 |  11 |\n",
       "| AAACGAAAGACACAAT-1 |       5 |       0 |     0 |      2 |      0 |     0 |       3 | None | 0 |   1 |    1 |   0 |   1 |    1 |  0 |   1 |\n",
       "| AAACGAAAGACACGGT-1 |    1643 |     417 |    11 |     12 |    214 |     8 |     981 | None | 0 |  97 |  502 | 100 |  17 |  540 |  5 |  82 |\n",
       "| AAACGAAAGACACTTC-1 |      11 |       3 |     0 |      2 |      0 |     0 |       6 | None | 0 |   1 |    4 |   2 |   0 |    4 |  0 |   1 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| TTTGTGTTCTGGGCGT-1 |     4 |    0 |   0 |   0 |    1 |  0 |     3 | None       | 0 |    0 |     2 |    1 |    0 |     2 |  0 |    1 |\n",
       "| TTTGTGTTCTGGTACA-1 |    56 |   14 |   0 |   1 |    1 |  0 |    40 | None       | 0 |   29 |    34 |    5 |   25 |    38 |  0 |   35 |\n",
       "| TTTGTGTTCTGTCGGG-1 |  1723 |  505 |  15 |  15 |  190 |  0 |   998 | None       | 0 |  117 |   462 |   89 |   25 |   514 | 12 |   76 |\n",
       "| TTTGTGTTCTGTCGTC-1 |    12 |    9 |   0 |   0 |    1 |  0 |     2 | None       | 0 |    1 |     2 |    1 |    1 |     2 |  0 |    2 |\n",
       "| TTTGTGTTCTGTGTCC-1 |     5 |    1 |   0 |   0 |    0 |  0 |     4 | None       | 0 |    1 |     1 |    0 |    1 |     1 |  0 |    1 |\n",
       "| TTTGTGTTCTGTGTGA-1 |  2460 |  678 |  17 |  24 |  338 |  2 |  1401 | None       | 0 |  134 |   652 |   99 |   33 |   703 | 11 |   92 |\n",
       "| TTTGTGTTCTGTTTAC-1 |     2 |    0 |   0 |   0 |    0 |  0 |     2 | None       | 0 |    0 |     1 |    0 |    0 |     1 |  0 |    0 |\n",
       "| TTTGTGTTCTTAAGCG-1 |     1 |    0 |   0 |   0 |    0 |  0 |     1 | None       | 0 |    1 |     0 |    0 |    0 |     1 |  0 |    0 |\n",
       "| TTTGTGTTCTTAATCC-1 |     2 |    0 |   0 |   0 |    0 |  0 |     2 | None       | 0 |    1 |     1 |    0 |    1 |     1 |  0 |    1 |\n",
       "| TTTGTGTTCTTACCTA-1 |     1 |    0 |   0 |   0 |    0 |  0 |     1 | None       | 0 |    0 |     0 |    0 |    0 |     0 |  0 |    0 |\n",
       "| TTTGTGTTCTTACGGA-1 |    10 |    1 |   0 |   1 |    1 |  1 |     6 | None       | 0 |    2 |     1 |    1 |    0 |     3 |  0 |    0 |\n",
       "| TTTGTGTTCTTACTCA-1 |     1 |    0 |   0 |   0 |    0 |  0 |     1 | None       | 0 |    0 |     1 |    1 |    0 |     1 |  0 |    1 |\n",
       "| TTTGTGTTCTTAGTGG-1 |     2 |    0 |   0 |   0 |    1 |  0 |     1 | None       | 0 |    1 |     1 |    1 |    0 |     1 |  0 |    1 |\n",
       "| TTTGTGTTCTTATCAC-1 |     3 |    2 |   0 |   0 |    0 |  0 |     1 | None       | 0 |    0 |     0 |    0 |    0 |     0 |  0 |    0 |\n",
       "| TTTGTGTTCTTCACTA-1 |     1 |    0 |   0 |   0 |    0 |  0 |     1 | None       | 0 |    0 |     1 |    0 |    0 |     1 |  0 |    0 |\n",
       "| TTTGTGTTCTTCAGAG-1 |     1 |    0 |   0 |   0 |    1 |  0 |     0 | None       | 0 |    0 |     0 |    0 |    0 |     0 |  0 |    0 |\n",
       "| TTTGTGTTCTTCATAC-1 |     4 |    0 |   0 |   0 |    0 |  0 |     4 | None       | 0 |    0 |     2 |    1 |    0 |     2 |  0 |    1 |\n",
       "| TTTGTGTTCTTCCACG-1 |     4 |    1 |   0 |   0 |    1 |  0 |     2 | None       | 0 |    2 |     2 |    0 |    2 |     2 |  0 |    2 |\n",
       "| TTTGTGTTCTTCCAGC-1 |     1 |    1 |   0 |   0 |    0 |  0 |     0 | None       | 0 |    0 |     0 |    0 |    0 |     0 |  0 |    0 |\n",
       "| TTTGTGTTCTTCCGTG-1 |    19 |    8 |   0 |   1 |    2 |  0 |     8 | None       | 0 |    4 |     5 |    0 |    2 |     6 |  0 |    3 |\n",
       "| TTTGTGTTCTTCGTTA-1 |    45 |   11 |   0 |   1 |    4 |  0 |    29 | None       | 0 |    9 |    20 |    4 |    8 |    20 |  0 |   16 |\n",
       "| TTTGTGTTCTTCTCTC-1 |     1 |    0 |   0 |   0 |    0 |  0 |     1 | None       | 0 |    0 |     0 |    0 |    0 |     0 |  0 |    0 |\n",
       "| TTTGTGTTCTTCTTCC-1 |    14 |    7 |   0 |   1 |    3 |  0 |     3 | None       | 0 |    2 |     3 |    0 |    2 |     3 |  0 |    2 |\n",
       "| TTTGTGTTCTTGCGCT-1 |    32 |    2 |   1 |   1 |    3 |  1 |    24 | None       | 0 |   12 |    17 |    2 |    9 |    18 |  0 |   12 |\n",
       "| TTTGTGTTCTTGGCCT-1 |     3 |    0 |   0 |   0 |    0 |  0 |     3 | None       | 0 |    0 |     3 |    1 |    0 |     3 |  0 |    1 |\n",
       "| TTTGTGTTCTTGTACT-1 |    47 |    2 |   0 |   0 |    4 |  0 |    41 | None       | 0 |   11 |    31 |    9 |    8 |    32 |  0 |   20 |\n",
       "| TTTGTGTTCTTGTCGC-1 |     1 |    0 |   0 |   0 |    0 |  0 |     1 | None       | 0 |    0 |     0 |    0 |    0 |     0 |  0 |    0 |\n",
       "| TTTGTGTTCTTGTGCC-1 | 23379 | 9625 | 160 | 115 | 1232 | 20 | 12227 | _cell_8727 | 1 | 6321 | 10005 | 2560 | 5242 | 10523 | 24 | 8309 |\n",
       "| TTTGTGTTCTTTATCG-1 |     2 |    1 |   0 |   0 |    0 |  0 |     1 | None       | 0 |    0 |     0 |    0 |    0 |     0 |  0 |    0 |\n",
       "| TTTGTGTTCTTTGCAT-1 |     3 |    1 |   0 |   0 |    0 |  0 |     2 | None       | 0 |    1 |     2 |    0 |    1 |     2 |  0 |    1 |\n",
       "\n"
      ],
      "text/plain": [
       "                   total   duplicate chimeric unmapped lowmapq mitochondrial\n",
       "NO_BARCODE         8335327 3922818   71559    746476   634014  22972        \n",
       "AAACGAAAGAAACGCC-1       3       1       0         1        0      0        \n",
       "AAACGAAAGAAAGCAG-1      14       1       0         4        2      0        \n",
       "AAACGAAAGAAAGGGT-1       7       1       0         1        0      0        \n",
       "AAACGAAAGAAATACC-1    9880    5380      79       106     1120      6        \n",
       "AAACGAAAGAAATCTG-1       2       0       0         0        0      0        \n",
       "AAACGAAAGAAATGGG-1       4       2       0         1        0      0        \n",
       "AAACGAAAGAAATTCG-1       2       0       0         0        1      0        \n",
       "AAACGAAAGAACCATA-1      12       3       0         8        0      0        \n",
       "AAACGAAAGAACCCGA-1       3       1       0         2        0      0        \n",
       "AAACGAAAGAACGACC-1       2       0       0         0        0      0        \n",
       "AAACGAAAGAACGCCA-1       7       0       0         1        0      0        \n",
       "AAACGAAAGAACGTCG-1       3       0       0         1        0      0        \n",
       "AAACGAAAGAACGTTA-1      11       1       0         2        2      0        \n",
       "AAACGAAAGAACTAAC-1       7       0       0         5        1      0        \n",
       "AAACGAAAGAACTCCT-1       2       2       0         0        0      0        \n",
       "AAACGAAAGAAGACTT-1       3       2       0         0        0      0        \n",
       "AAACGAAAGAAGAGTG-1       4       1       0         0        0      0        \n",
       "AAACGAAAGAAGCCGT-1       2       0       0         0        1      0        \n",
       "AAACGAAAGAAGCCTG-1       7       0       0         0        1      0        \n",
       "AAACGAAAGAAGGGCG-1      10       0       2         1        1      0        \n",
       "AAACGAAAGAATACTG-1      10       1       0         2        2      0        \n",
       "AAACGAAAGAATAGTC-1     464      79       0        10       40      0        \n",
       "AAACGAAAGAATCAAC-1      11       2       0         4        1      0        \n",
       "AAACGAAAGAATCAGT-1       4       1       0         1        0      0        \n",
       "AAACGAAAGAATGCTA-1       5       2       0         0        2      0        \n",
       "AAACGAAAGACAACGC-1      30       6       0         6        0      0        \n",
       "AAACGAAAGACACAAT-1       5       0       0         2        0      0        \n",
       "AAACGAAAGACACGGT-1    1643     417      11        12      214      8        \n",
       "AAACGAAAGACACTTC-1      11       3       0         2        0      0        \n",
       "⋮                  ⋮       ⋮         ⋮        ⋮        ⋮       ⋮            \n",
       "TTTGTGTTCTGGGCGT-1     4      0        0        0         1     0           \n",
       "TTTGTGTTCTGGTACA-1    56     14        0        1         1     0           \n",
       "TTTGTGTTCTGTCGGG-1  1723    505       15       15       190     0           \n",
       "TTTGTGTTCTGTCGTC-1    12      9        0        0         1     0           \n",
       "TTTGTGTTCTGTGTCC-1     5      1        0        0         0     0           \n",
       "TTTGTGTTCTGTGTGA-1  2460    678       17       24       338     2           \n",
       "TTTGTGTTCTGTTTAC-1     2      0        0        0         0     0           \n",
       "TTTGTGTTCTTAAGCG-1     1      0        0        0         0     0           \n",
       "TTTGTGTTCTTAATCC-1     2      0        0        0         0     0           \n",
       "TTTGTGTTCTTACCTA-1     1      0        0        0         0     0           \n",
       "TTTGTGTTCTTACGGA-1    10      1        0        1         1     1           \n",
       "TTTGTGTTCTTACTCA-1     1      0        0        0         0     0           \n",
       "TTTGTGTTCTTAGTGG-1     2      0        0        0         1     0           \n",
       "TTTGTGTTCTTATCAC-1     3      2        0        0         0     0           \n",
       "TTTGTGTTCTTCACTA-1     1      0        0        0         0     0           \n",
       "TTTGTGTTCTTCAGAG-1     1      0        0        0         1     0           \n",
       "TTTGTGTTCTTCATAC-1     4      0        0        0         0     0           \n",
       "TTTGTGTTCTTCCACG-1     4      1        0        0         1     0           \n",
       "TTTGTGTTCTTCCAGC-1     1      1        0        0         0     0           \n",
       "TTTGTGTTCTTCCGTG-1    19      8        0        1         2     0           \n",
       "TTTGTGTTCTTCGTTA-1    45     11        0        1         4     0           \n",
       "TTTGTGTTCTTCTCTC-1     1      0        0        0         0     0           \n",
       "TTTGTGTTCTTCTTCC-1    14      7        0        1         3     0           \n",
       "TTTGTGTTCTTGCGCT-1    32      2        1        1         3     1           \n",
       "TTTGTGTTCTTGGCCT-1     3      0        0        0         0     0           \n",
       "TTTGTGTTCTTGTACT-1    47      2        0        0         4     0           \n",
       "TTTGTGTTCTTGTCGC-1     1      0        0        0         0     0           \n",
       "TTTGTGTTCTTGTGCC-1 23379   9625      160      115      1232    20           \n",
       "TTTGTGTTCTTTATCG-1     2      1        0        0         0     0           \n",
       "TTTGTGTTCTTTGCAT-1     3      1        0        0         0     0           \n",
       "                   passed_filters cell_id    is__cell_barcode TSS_fragments\n",
       "NO_BARCODE         2937488        None       0                  0          \n",
       "AAACGAAAGAAACGCC-1       1        None       0                  1          \n",
       "AAACGAAAGAAAGCAG-1       7        None       0                  0          \n",
       "AAACGAAAGAAAGGGT-1       5        None       0                  2          \n",
       "AAACGAAAGAAATACC-1    3189        None       0                386          \n",
       "AAACGAAAGAAATCTG-1       2        None       0                  1          \n",
       "AAACGAAAGAAATGGG-1       1        None       0                  0          \n",
       "AAACGAAAGAAATTCG-1       1        None       0                  0          \n",
       "AAACGAAAGAACCATA-1       1        None       0                  0          \n",
       "AAACGAAAGAACCCGA-1       0        None       0                  0          \n",
       "AAACGAAAGAACGACC-1       2        None       0                  2          \n",
       "AAACGAAAGAACGCCA-1       6        None       0                  0          \n",
       "AAACGAAAGAACGTCG-1       2        None       0                  0          \n",
       "AAACGAAAGAACGTTA-1       6        None       0                  2          \n",
       "AAACGAAAGAACTAAC-1       1        None       0                  1          \n",
       "AAACGAAAGAACTCCT-1       0        None       0                  0          \n",
       "AAACGAAAGAAGACTT-1       1        None       0                  0          \n",
       "AAACGAAAGAAGAGTG-1       3        None       0                  0          \n",
       "AAACGAAAGAAGCCGT-1       1        None       0                  0          \n",
       "AAACGAAAGAAGCCTG-1       6        None       0                  0          \n",
       "AAACGAAAGAAGGGCG-1       6        None       0                  0          \n",
       "AAACGAAAGAATACTG-1       5        None       0                  0          \n",
       "AAACGAAAGAATAGTC-1     335        None       0                 43          \n",
       "AAACGAAAGAATCAAC-1       4        None       0                  0          \n",
       "AAACGAAAGAATCAGT-1       2        None       0                  0          \n",
       "AAACGAAAGAATGCTA-1       1        None       0                  0          \n",
       "AAACGAAAGACAACGC-1      18        None       0                  8          \n",
       "AAACGAAAGACACAAT-1       3        None       0                  1          \n",
       "AAACGAAAGACACGGT-1     981        None       0                 97          \n",
       "AAACGAAAGACACTTC-1       6        None       0                  1          \n",
       "⋮                  ⋮              ⋮          ⋮                ⋮            \n",
       "TTTGTGTTCTGGGCGT-1     3          None       0                   0         \n",
       "TTTGTGTTCTGGTACA-1    40          None       0                  29         \n",
       "TTTGTGTTCTGTCGGG-1   998          None       0                 117         \n",
       "TTTGTGTTCTGTCGTC-1     2          None       0                   1         \n",
       "TTTGTGTTCTGTGTCC-1     4          None       0                   1         \n",
       "TTTGTGTTCTGTGTGA-1  1401          None       0                 134         \n",
       "TTTGTGTTCTGTTTAC-1     2          None       0                   0         \n",
       "TTTGTGTTCTTAAGCG-1     1          None       0                   1         \n",
       "TTTGTGTTCTTAATCC-1     2          None       0                   1         \n",
       "TTTGTGTTCTTACCTA-1     1          None       0                   0         \n",
       "TTTGTGTTCTTACGGA-1     6          None       0                   2         \n",
       "TTTGTGTTCTTACTCA-1     1          None       0                   0         \n",
       "TTTGTGTTCTTAGTGG-1     1          None       0                   1         \n",
       "TTTGTGTTCTTATCAC-1     1          None       0                   0         \n",
       "TTTGTGTTCTTCACTA-1     1          None       0                   0         \n",
       "TTTGTGTTCTTCAGAG-1     0          None       0                   0         \n",
       "TTTGTGTTCTTCATAC-1     4          None       0                   0         \n",
       "TTTGTGTTCTTCCACG-1     2          None       0                   2         \n",
       "TTTGTGTTCTTCCAGC-1     0          None       0                   0         \n",
       "TTTGTGTTCTTCCGTG-1     8          None       0                   4         \n",
       "TTTGTGTTCTTCGTTA-1    29          None       0                   9         \n",
       "TTTGTGTTCTTCTCTC-1     1          None       0                   0         \n",
       "TTTGTGTTCTTCTTCC-1     3          None       0                   2         \n",
       "TTTGTGTTCTTGCGCT-1    24          None       0                  12         \n",
       "TTTGTGTTCTTGGCCT-1     3          None       0                   0         \n",
       "TTTGTGTTCTTGTACT-1    41          None       0                  11         \n",
       "TTTGTGTTCTTGTCGC-1     1          None       0                   0         \n",
       "TTTGTGTTCTTGTGCC-1 12227          _cell_8727 1                6321         \n",
       "TTTGTGTTCTTTATCG-1     1          None       0                   0         \n",
       "TTTGTGTTCTTTGCAT-1     2          None       0                   1         \n",
       "                   DNase_sensitive_region_fragments enhancer_region_fragments\n",
       "NO_BARCODE            0                               0                      \n",
       "AAACGAAAGAAACGCC-1    1                               1                      \n",
       "AAACGAAAGAAAGCAG-1    2                               0                      \n",
       "AAACGAAAGAAAGGGT-1    2                               0                      \n",
       "AAACGAAAGAAATACC-1 1623                             297                      \n",
       "AAACGAAAGAAATCTG-1    2                               1                      \n",
       "AAACGAAAGAAATGGG-1    1                               1                      \n",
       "AAACGAAAGAAATTCG-1    0                               0                      \n",
       "AAACGAAAGAACCATA-1    1                               0                      \n",
       "AAACGAAAGAACCCGA-1    0                               0                      \n",
       "AAACGAAAGAACGACC-1    2                               1                      \n",
       "AAACGAAAGAACGCCA-1    4                               0                      \n",
       "AAACGAAAGAACGTCG-1    1                               1                      \n",
       "AAACGAAAGAACGTTA-1    6                               1                      \n",
       "AAACGAAAGAACTAAC-1    1                               0                      \n",
       "AAACGAAAGAACTCCT-1    0                               0                      \n",
       "AAACGAAAGAAGACTT-1    1                               1                      \n",
       "AAACGAAAGAAGAGTG-1    1                               1                      \n",
       "AAACGAAAGAAGCCGT-1    0                               0                      \n",
       "AAACGAAAGAAGCCTG-1    4                               0                      \n",
       "AAACGAAAGAAGGGCG-1    3                               1                      \n",
       "AAACGAAAGAATACTG-1    4                               0                      \n",
       "AAACGAAAGAATAGTC-1  176                              31                      \n",
       "AAACGAAAGAATCAAC-1    0                               1                      \n",
       "AAACGAAAGAATCAGT-1    1                               1                      \n",
       "AAACGAAAGAATGCTA-1    0                               0                      \n",
       "AAACGAAAGACAACGC-1   15                               2                      \n",
       "AAACGAAAGACACAAT-1    1                               0                      \n",
       "AAACGAAAGACACGGT-1  502                             100                      \n",
       "AAACGAAAGACACTTC-1    4                               2                      \n",
       "⋮                  ⋮                                ⋮                        \n",
       "TTTGTGTTCTGGGCGT-1     2                               1                     \n",
       "TTTGTGTTCTGGTACA-1    34                               5                     \n",
       "TTTGTGTTCTGTCGGG-1   462                              89                     \n",
       "TTTGTGTTCTGTCGTC-1     2                               1                     \n",
       "TTTGTGTTCTGTGTCC-1     1                               0                     \n",
       "TTTGTGTTCTGTGTGA-1   652                              99                     \n",
       "TTTGTGTTCTGTTTAC-1     1                               0                     \n",
       "TTTGTGTTCTTAAGCG-1     0                               0                     \n",
       "TTTGTGTTCTTAATCC-1     1                               0                     \n",
       "TTTGTGTTCTTACCTA-1     0                               0                     \n",
       "TTTGTGTTCTTACGGA-1     1                               1                     \n",
       "TTTGTGTTCTTACTCA-1     1                               1                     \n",
       "TTTGTGTTCTTAGTGG-1     1                               1                     \n",
       "TTTGTGTTCTTATCAC-1     0                               0                     \n",
       "TTTGTGTTCTTCACTA-1     1                               0                     \n",
       "TTTGTGTTCTTCAGAG-1     0                               0                     \n",
       "TTTGTGTTCTTCATAC-1     2                               1                     \n",
       "TTTGTGTTCTTCCACG-1     2                               0                     \n",
       "TTTGTGTTCTTCCAGC-1     0                               0                     \n",
       "TTTGTGTTCTTCCGTG-1     5                               0                     \n",
       "TTTGTGTTCTTCGTTA-1    20                               4                     \n",
       "TTTGTGTTCTTCTCTC-1     0                               0                     \n",
       "TTTGTGTTCTTCTTCC-1     3                               0                     \n",
       "TTTGTGTTCTTGCGCT-1    17                               2                     \n",
       "TTTGTGTTCTTGGCCT-1     3                               1                     \n",
       "TTTGTGTTCTTGTACT-1    31                               9                     \n",
       "TTTGTGTTCTTGTCGC-1     0                               0                     \n",
       "TTTGTGTTCTTGTGCC-1 10005                            2560                     \n",
       "TTTGTGTTCTTTATCG-1     0                               0                     \n",
       "TTTGTGTTCTTTGCAT-1     2                               0                     \n",
       "                   promoter_region_fragments on_target_fragments\n",
       "NO_BARCODE           0                          0               \n",
       "AAACGAAAGAAACGCC-1   0                          1               \n",
       "AAACGAAAGAAAGCAG-1   0                          2               \n",
       "AAACGAAAGAAAGGGT-1   1                          2               \n",
       "AAACGAAAGAAATACC-1 107                       1758               \n",
       "AAACGAAAGAAATCTG-1   1                          2               \n",
       "AAACGAAAGAAATGGG-1   0                          1               \n",
       "AAACGAAAGAAATTCG-1   0                          0               \n",
       "AAACGAAAGAACCATA-1   0                          1               \n",
       "AAACGAAAGAACCCGA-1   0                          0               \n",
       "AAACGAAAGAACGACC-1   1                          2               \n",
       "AAACGAAAGAACGCCA-1   0                          4               \n",
       "AAACGAAAGAACGTCG-1   0                          1               \n",
       "AAACGAAAGAACGTTA-1   2                          6               \n",
       "AAACGAAAGAACTAAC-1   1                          1               \n",
       "AAACGAAAGAACTCCT-1   0                          0               \n",
       "AAACGAAAGAAGACTT-1   0                          1               \n",
       "AAACGAAAGAAGAGTG-1   0                          2               \n",
       "AAACGAAAGAAGCCGT-1   0                          0               \n",
       "AAACGAAAGAAGCCTG-1   0                          4               \n",
       "AAACGAAAGAAGGGCG-1   0                          3               \n",
       "AAACGAAAGAATACTG-1   0                          4               \n",
       "AAACGAAAGAATAGTC-1   4                        194               \n",
       "AAACGAAAGAATCAAC-1   0                          1               \n",
       "AAACGAAAGAATCAGT-1   0                          1               \n",
       "AAACGAAAGAATGCTA-1   0                          0               \n",
       "AAACGAAAGACAACGC-1   7                         16               \n",
       "AAACGAAAGACACAAT-1   1                          1               \n",
       "AAACGAAAGACACGGT-1  17                        540               \n",
       "AAACGAAAGACACTTC-1   0                          4               \n",
       "⋮                  ⋮                         ⋮                  \n",
       "TTTGTGTTCTGGGCGT-1    0                          2              \n",
       "TTTGTGTTCTGGTACA-1   25                         38              \n",
       "TTTGTGTTCTGTCGGG-1   25                        514              \n",
       "TTTGTGTTCTGTCGTC-1    1                          2              \n",
       "TTTGTGTTCTGTGTCC-1    1                          1              \n",
       "TTTGTGTTCTGTGTGA-1   33                        703              \n",
       "TTTGTGTTCTGTTTAC-1    0                          1              \n",
       "TTTGTGTTCTTAAGCG-1    0                          1              \n",
       "TTTGTGTTCTTAATCC-1    1                          1              \n",
       "TTTGTGTTCTTACCTA-1    0                          0              \n",
       "TTTGTGTTCTTACGGA-1    0                          3              \n",
       "TTTGTGTTCTTACTCA-1    0                          1              \n",
       "TTTGTGTTCTTAGTGG-1    0                          1              \n",
       "TTTGTGTTCTTATCAC-1    0                          0              \n",
       "TTTGTGTTCTTCACTA-1    0                          1              \n",
       "TTTGTGTTCTTCAGAG-1    0                          0              \n",
       "TTTGTGTTCTTCATAC-1    0                          2              \n",
       "TTTGTGTTCTTCCACG-1    2                          2              \n",
       "TTTGTGTTCTTCCAGC-1    0                          0              \n",
       "TTTGTGTTCTTCCGTG-1    2                          6              \n",
       "TTTGTGTTCTTCGTTA-1    8                         20              \n",
       "TTTGTGTTCTTCTCTC-1    0                          0              \n",
       "TTTGTGTTCTTCTTCC-1    2                          3              \n",
       "TTTGTGTTCTTGCGCT-1    9                         18              \n",
       "TTTGTGTTCTTGGCCT-1    0                          3              \n",
       "TTTGTGTTCTTGTACT-1    8                         32              \n",
       "TTTGTGTTCTTGTCGC-1    0                          0              \n",
       "TTTGTGTTCTTGTGCC-1 5242                      10523              \n",
       "TTTGTGTTCTTTATCG-1    0                          0              \n",
       "TTTGTGTTCTTTGCAT-1    1                          2              \n",
       "                   blacklist_region_fragments peak_region_fragments\n",
       "NO_BARCODE          0                           0                  \n",
       "AAACGAAAGAAACGCC-1  0                           1                  \n",
       "AAACGAAAGAAAGCAG-1  0                           0                  \n",
       "AAACGAAAGAAAGGGT-1  0                           1                  \n",
       "AAACGAAAGAAATACC-1 16                         262                  \n",
       "AAACGAAAGAAATCTG-1  0                           2                  \n",
       "AAACGAAAGAAATGGG-1  0                           1                  \n",
       "AAACGAAAGAAATTCG-1  0                           0                  \n",
       "AAACGAAAGAACCATA-1  0                           0                  \n",
       "AAACGAAAGAACCCGA-1  0                           0                  \n",
       "AAACGAAAGAACGACC-1  0                           1                  \n",
       "AAACGAAAGAACGCCA-1  0                           0                  \n",
       "AAACGAAAGAACGTCG-1  0                           1                  \n",
       "AAACGAAAGAACGTTA-1  0                           4                  \n",
       "AAACGAAAGAACTAAC-1  0                           1                  \n",
       "AAACGAAAGAACTCCT-1  0                           0                  \n",
       "AAACGAAAGAAGACTT-1  0                           1                  \n",
       "AAACGAAAGAAGAGTG-1  0                           0                  \n",
       "AAACGAAAGAAGCCGT-1  0                           0                  \n",
       "AAACGAAAGAAGCCTG-1  0                           0                  \n",
       "AAACGAAAGAAGGGCG-1  0                           2                  \n",
       "AAACGAAAGAATACTG-1  0                           0                  \n",
       "AAACGAAAGAATAGTC-1  2                          19                  \n",
       "AAACGAAAGAATCAAC-1  0                           1                  \n",
       "AAACGAAAGAATCAGT-1  0                           1                  \n",
       "AAACGAAAGAATGCTA-1  0                           0                  \n",
       "AAACGAAAGACAACGC-1  0                          11                  \n",
       "AAACGAAAGACACAAT-1  0                           1                  \n",
       "AAACGAAAGACACGGT-1  5                          82                  \n",
       "AAACGAAAGACACTTC-1  0                           1                  \n",
       "⋮                  ⋮                          ⋮                    \n",
       "TTTGTGTTCTGGGCGT-1  0                            1                 \n",
       "TTTGTGTTCTGGTACA-1  0                           35                 \n",
       "TTTGTGTTCTGTCGGG-1 12                           76                 \n",
       "TTTGTGTTCTGTCGTC-1  0                            2                 \n",
       "TTTGTGTTCTGTGTCC-1  0                            1                 \n",
       "TTTGTGTTCTGTGTGA-1 11                           92                 \n",
       "TTTGTGTTCTGTTTAC-1  0                            0                 \n",
       "TTTGTGTTCTTAAGCG-1  0                            0                 \n",
       "TTTGTGTTCTTAATCC-1  0                            1                 \n",
       "TTTGTGTTCTTACCTA-1  0                            0                 \n",
       "TTTGTGTTCTTACGGA-1  0                            0                 \n",
       "TTTGTGTTCTTACTCA-1  0                            1                 \n",
       "TTTGTGTTCTTAGTGG-1  0                            1                 \n",
       "TTTGTGTTCTTATCAC-1  0                            0                 \n",
       "TTTGTGTTCTTCACTA-1  0                            0                 \n",
       "TTTGTGTTCTTCAGAG-1  0                            0                 \n",
       "TTTGTGTTCTTCATAC-1  0                            1                 \n",
       "TTTGTGTTCTTCCACG-1  0                            2                 \n",
       "TTTGTGTTCTTCCAGC-1  0                            0                 \n",
       "TTTGTGTTCTTCCGTG-1  0                            3                 \n",
       "TTTGTGTTCTTCGTTA-1  0                           16                 \n",
       "TTTGTGTTCTTCTCTC-1  0                            0                 \n",
       "TTTGTGTTCTTCTTCC-1  0                            2                 \n",
       "TTTGTGTTCTTGCGCT-1  0                           12                 \n",
       "TTTGTGTTCTTGGCCT-1  0                            1                 \n",
       "TTTGTGTTCTTGTACT-1  0                           20                 \n",
       "TTTGTGTTCTTGTCGC-1  0                            0                 \n",
       "TTTGTGTTCTTGTGCC-1 24                         8309                 \n",
       "TTTGTGTTCTTTATCG-1  0                            0                 \n",
       "TTTGTGTTCTTTGCAT-1  0                            1                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata <- read.csv(file = '../data/atac/atac_v1_pbmc_10k_singlecell.csv', header = T, row.names = 1)\n",
    "View(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2ec187bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc <- CreateSeuratObject(\n",
    "  counts = chrom_assay,\n",
    "  meta.data = metadata,\n",
    "  assay = 'ATAC'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "682289bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'Seurat' [package \"SeuratObject\"] with 13 slots\n",
      "  ..@ assays      :List of 1\n",
      "  .. ..$ ATAC:Formal class 'ChromatinAssay' [package \"Signac\"] with 16 slots\n",
      "  .. .. .. ..@ ranges            :Formal class 'GRanges' [package \"GenomicRanges\"] with 7 slots\n",
      "  .. .. .. .. .. ..@ seqnames       :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. .. .. .. .. .. ..@ values         : Factor w/ 24 levels \"chr1\",\"chr2\",..: 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. .. .. .. ..@ lengths        : int [1:24] 8555 6858 5499 3515 4389 5269 4263 3677 3540 4137 ...\n",
      "  .. .. .. .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. .. .. .. ..@ ranges         :Formal class 'IRanges' [package \"IRanges\"] with 6 slots\n",
      "  .. .. .. .. .. .. .. ..@ start          : int [1:87561] 565107 569174 713460 752422 762106 779589 793516 801120 804872 839520 ...\n",
      "  .. .. .. .. .. .. .. ..@ width          : int [1:87561] 444 466 1364 617 1254 683 226 219 890 1604 ...\n",
      "  .. .. .. .. .. .. .. ..@ NAMES          : NULL\n",
      "  .. .. .. .. .. .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. .. .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. .. .. .. ..@ strand         :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. .. .. .. .. .. ..@ values         : Factor w/ 3 levels \"+\",\"-\",\"*\": 3\n",
      "  .. .. .. .. .. .. .. ..@ lengths        : int 87561\n",
      "  .. .. .. .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. .. .. .. ..@ seqinfo        :Formal class 'Seqinfo' [package \"GenomeInfoDb\"] with 4 slots\n",
      "  .. .. .. .. .. .. .. ..@ seqnames   : chr [1:24] \"chr1\" \"chr2\" \"chr3\" \"chr4\" ...\n",
      "  .. .. .. .. .. .. .. ..@ seqlengths : int [1:24] NA NA NA NA NA NA NA NA NA NA ...\n",
      "  .. .. .. .. .. .. .. ..@ is_circular: logi [1:24] NA NA NA NA NA NA ...\n",
      "  .. .. .. .. .. .. .. ..@ genome     : chr [1:24] NA NA NA NA ...\n",
      "  .. .. .. .. .. ..@ elementMetadata:Formal class 'DFrame' [package \"S4Vectors\"] with 6 slots\n",
      "  .. .. .. .. .. .. .. ..@ rownames       : NULL\n",
      "  .. .. .. .. .. .. .. ..@ nrows          : int 87561\n",
      "  .. .. .. .. .. .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. .. .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. .. .. .. .. .. ..@ listData       : Named list()\n",
      "  .. .. .. .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. .. ..@ motifs            : NULL\n",
      "  .. .. .. ..@ fragments         :List of 1\n",
      "  .. .. .. .. ..$ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. .. .. .. .. .. ..@ path : chr \"/home/kappa/git/goldfish/data/atac/atac_v1_pbmc_10k_fragments.tsv.gz\"\n",
      "  .. .. .. .. .. .. ..@ hash : chr [1:2] \"8e03bd72e952a7da7c1cce52f3696e6e\" \"146e556e1311573593bd6cfa6efefcc1\"\n",
      "  .. .. .. .. .. .. ..@ cells: Named chr [1:8728] \"AAACGAAAGAGCGAAA-1\" \"AAACGAAAGAGTTTGA-1\" \"AAACGAAAGCGAGCTA-1\" \"AAACGAAAGGCTTCGC-1\" ...\n",
      "  .. .. .. .. .. .. .. ..- attr(*, \"names\")= chr [1:8728] \"AAACGAAAGAGCGAAA-1\" \"AAACGAAAGAGTTTGA-1\" \"AAACGAAAGCGAGCTA-1\" \"AAACGAAAGGCTTCGC-1\" ...\n",
      "  .. .. .. ..@ seqinfo           : NULL\n",
      "  .. .. .. ..@ annotation        : NULL\n",
      "  .. .. .. ..@ bias              : NULL\n",
      "  .. .. .. ..@ positionEnrichment: list()\n",
      "  .. .. .. ..@ links             :Formal class 'GRanges' [package \"GenomicRanges\"] with 7 slots\n",
      "  .. .. .. .. .. ..@ seqnames       :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. .. .. .. .. .. ..@ values         : Factor w/ 0 levels: \n",
      "  .. .. .. .. .. .. .. ..@ lengths        : int(0) \n",
      "  .. .. .. .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. .. .. .. ..@ ranges         :Formal class 'IRanges' [package \"IRanges\"] with 6 slots\n",
      "  .. .. .. .. .. .. .. ..@ start          : int(0) \n",
      "  .. .. .. .. .. .. .. ..@ width          : int(0) \n",
      "  .. .. .. .. .. .. .. ..@ NAMES          : NULL\n",
      "  .. .. .. .. .. .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. .. .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. .. .. .. ..@ strand         :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. .. .. .. .. .. ..@ values         : Factor w/ 3 levels \"+\",\"-\",\"*\": \n",
      "  .. .. .. .. .. .. .. ..@ lengths        : int(0) \n",
      "  .. .. .. .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. .. .. .. ..@ seqinfo        :Formal class 'Seqinfo' [package \"GenomeInfoDb\"] with 4 slots\n",
      "  .. .. .. .. .. .. .. ..@ seqnames   : chr(0) \n",
      "  .. .. .. .. .. .. .. ..@ seqlengths : int(0) \n",
      "  .. .. .. .. .. .. .. ..@ is_circular: logi(0) \n",
      "  .. .. .. .. .. .. .. ..@ genome     : chr(0) \n",
      "  .. .. .. .. .. ..@ elementMetadata:Formal class 'DFrame' [package \"S4Vectors\"] with 6 slots\n",
      "  .. .. .. .. .. .. .. ..@ rownames       : NULL\n",
      "  .. .. .. .. .. .. .. ..@ nrows          : int 0\n",
      "  .. .. .. .. .. .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. .. .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. .. .. .. .. .. ..@ listData       : Named list()\n",
      "  .. .. .. .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. .. ..@ counts            :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int [1:52681465] 31 74 83 116 134 147 153 166 175 178 ...\n",
      "  .. .. .. .. .. ..@ p       : int [1:8729] 0 5546 12127 21340 51653 56574 61650 64340 73157 79801 ...\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 87561 8728\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:87561] \"chr1-565107-565550\" \"chr1-569174-569639\" \"chr1-713460-714823\" \"chr1-752422-753038\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr [1:8728] \"AAACGAAAGAGCGAAA-1\" \"AAACGAAAGAGTTTGA-1\" \"AAACGAAAGCGAGCTA-1\" \"AAACGAAAGGCTTCGC-1\" ...\n",
      "  .. .. .. .. .. ..@ x       : num [1:52681465] 2 2 2 1 2 2 2 6 2 2 ...\n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ data              :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int [1:52681465] 31 74 83 116 134 147 153 166 175 178 ...\n",
      "  .. .. .. .. .. ..@ p       : int [1:8729] 0 5546 12127 21340 51653 56574 61650 64340 73157 79801 ...\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 87561 8728\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:87561] \"chr1-565107-565550\" \"chr1-569174-569639\" \"chr1-713460-714823\" \"chr1-752422-753038\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr [1:8728] \"AAACGAAAGAGCGAAA-1\" \"AAACGAAAGAGTTTGA-1\" \"AAACGAAAGCGAGCTA-1\" \"AAACGAAAGGCTTCGC-1\" ...\n",
      "  .. .. .. .. .. ..@ x       : num [1:52681465] 2 2 2 1 2 2 2 6 2 2 ...\n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ scale.data        : num[0 , 0 ] \n",
      "  .. .. .. ..@ assay.orig        : NULL\n",
      "  .. .. .. ..@ var.features      : logi(0) \n",
      "  .. .. .. ..@ meta.features     :'data.frame':\t87561 obs. of  0 variables\n",
      "  .. .. .. ..@ misc              :List of 1\n",
      "  .. .. .. .. ..$ calcN: logi TRUE\n",
      "  .. .. .. ..@ key               : chr \"atac_\"\n",
      "  ..@ meta.data   :'data.frame':\t8728 obs. of  19 variables:\n",
      "  .. ..$ orig.ident                      : Factor w/ 1 level \"SeuratProject\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ nCount_ATAC                     : num [1:8728] 13187 16069 28008 221734 11439 ...\n",
      "  .. ..$ nFeature_ATAC                   : int [1:8728] 5546 6581 9213 30313 4921 5076 2690 8817 6644 3632 ...\n",
      "  .. ..$ total                           : int [1:8728] 22965 21806 46233 288897 16103 266481 10753 31745 23136 14626 ...\n",
      "  .. ..$ duplicate                       : int [1:8728] 10872 9336 21197 99998 6751 45654 4760 14341 10210 5612 ...\n",
      "  .. ..$ chimeric                        : int [1:8728] 183 122 479 2830 104 3425 61 167 145 99 ...\n",
      "  .. ..$ unmapped                        : int [1:8728] 198 175 292 1512 120 1514 43 191 157 109 ...\n",
      "  .. ..$ lowmapq                         : int [1:8728] 1198 1050 2673 16347 867 41940 644 1465 1308 913 ...\n",
      "  .. ..$ mitochondrial                   : int [1:8728] 51 1 14 913 0 57 96 29 15 96 ...\n",
      "  .. ..$ passed_filters                  : int [1:8728] 10463 11122 21578 167297 8261 173891 5149 15552 11301 7797 ...\n",
      "  .. ..$ cell_id                         : chr [1:8728] \"_cell_0\" \"_cell_1\" \"_cell_2\" \"_cell_3\" ...\n",
      "  .. ..$ is__cell_barcode                : int [1:8728] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ TSS_fragments                   : int [1:8728] 4157 4858 10696 86433 3784 14141 2637 6587 5136 3493 ...\n",
      "  .. ..$ DNase_sensitive_region_fragments: int [1:8728] 8537 9548 17420 137387 6973 78696 3973 13546 9636 6050 ...\n",
      "  .. ..$ enhancer_region_fragments       : int [1:8728] 3079 3954 4332 32884 2509 11425 753 5745 3542 1394 ...\n",
      "  .. ..$ promoter_region_fragments       : int [1:8728] 3249 3600 8763 70515 2987 1686 2289 4911 4019 2740 ...\n",
      "  .. ..$ on_target_fragments             : int [1:8728] 8894 10034 18562 144918 7288 85168 4183 14115 10072 6346 ...\n",
      "  .. ..$ blacklist_region_fragments      : int [1:8728] 6 10 22 300 4 2121 8 23 8 13 ...\n",
      "  .. ..$ peak_region_fragments           : int [1:8728] 6796 8319 14333 113658 5879 7360 3220 11399 8215 4509 ...\n",
      "  ..@ active.assay: chr \"ATAC\"\n",
      "  ..@ active.ident: Factor w/ 1 level \"SeuratProject\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..- attr(*, \"names\")= chr [1:8728] \"AAACGAAAGAGCGAAA-1\" \"AAACGAAAGAGTTTGA-1\" \"AAACGAAAGCGAGCTA-1\" \"AAACGAAAGGCTTCGC-1\" ...\n",
      "  ..@ graphs      : list()\n",
      "  ..@ neighbors   : list()\n",
      "  ..@ reductions  : list()\n",
      "  ..@ images      : list()\n",
      "  ..@ project.name: chr \"SeuratProject\"\n",
      "  ..@ misc        : list()\n",
      "  ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1\n",
      "  .. ..$ : int [1:3] 5 0 2\n",
      "  ..@ commands    : list()\n",
      "  ..@ tools       : list()\n"
     ]
    }
   ],
   "source": [
    "str(pbmc)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a93eed3",
   "metadata": {},
   "source": [
    "We want to store annotations into pbmc Serat object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c1222389",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NULL"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pbmc@assays$ATAC@annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15199249",
   "metadata": {},
   "source": [
    "When we look at the `annotations`, we see that they are not compatible with 10x genomics pbmc data we have. We need to add `chr` in front of each seqnames column. This is because the data was mapped to `hg19`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b6d91464",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n"
     ]
    }
   ],
   "source": [
    "annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4eeb569f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3072120 ranges and 5 metadata columns:\n",
       "                  seqnames        ranges strand |           tx_id   gene_name\n",
       "                     <Rle>     <IRanges>  <Rle> |     <character> <character>\n",
       "  ENSE00001489430        X 192989-193061      + | ENST00000399012      PLCXD1\n",
       "  ENSE00001536003        X 192991-193061      + | ENST00000484611      PLCXD1\n",
       "  ENSE00002160563        X 193020-193061      + | ENST00000430923      PLCXD1\n",
       "  ENSE00001750899        X 197722-197788      + | ENST00000445062      PLCXD1\n",
       "  ENSE00001489388        X 197859-198351      + | ENST00000381657      PLCXD1\n",
       "              ...      ...           ...    ... .             ...         ...\n",
       "  ENST00000361739       MT     7586-8269      + | ENST00000361739      MT-CO2\n",
       "  ENST00000361789       MT   14747-15887      + | ENST00000361789      MT-CYB\n",
       "  ENST00000361851       MT     8366-8572      + | ENST00000361851     MT-ATP8\n",
       "  ENST00000361899       MT     8527-9207      + | ENST00000361899     MT-ATP6\n",
       "  ENST00000362079       MT     9207-9990      + | ENST00000362079      MT-CO3\n",
       "                          gene_id   gene_biotype     type\n",
       "                      <character>    <character> <factor>\n",
       "  ENSE00001489430 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00001536003 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00002160563 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00001750899 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00001489388 ENSG00000182378 protein_coding     exon\n",
       "              ...             ...            ...      ...\n",
       "  ENST00000361739 ENSG00000198712 protein_coding      cds\n",
       "  ENST00000361789 ENSG00000198727 protein_coding      cds\n",
       "  ENST00000361851 ENSG00000228253 protein_coding      cds\n",
       "  ENST00000361899 ENSG00000198899 protein_coding      cds\n",
       "  ENST00000362079 ENSG00000198938 protein_coding      cds\n",
       "  -------\n",
       "  seqinfo: 25 sequences (1 circular) from GRCh37 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b90fb0a4",
   "metadata": {},
   "source": [
    "Add the gene information to the object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "181fe7ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3072120 ranges and 5 metadata columns:\n",
       "                  seqnames        ranges strand |           tx_id   gene_name\n",
       "                     <Rle>     <IRanges>  <Rle> |     <character> <character>\n",
       "  ENSE00001489430        X 192989-193061      + | ENST00000399012      PLCXD1\n",
       "  ENSE00001536003        X 192991-193061      + | ENST00000484611      PLCXD1\n",
       "  ENSE00002160563        X 193020-193061      + | ENST00000430923      PLCXD1\n",
       "  ENSE00001750899        X 197722-197788      + | ENST00000445062      PLCXD1\n",
       "  ENSE00001489388        X 197859-198351      + | ENST00000381657      PLCXD1\n",
       "              ...      ...           ...    ... .             ...         ...\n",
       "  ENST00000361739       MT     7586-8269      + | ENST00000361739      MT-CO2\n",
       "  ENST00000361789       MT   14747-15887      + | ENST00000361789      MT-CYB\n",
       "  ENST00000361851       MT     8366-8572      + | ENST00000361851     MT-ATP8\n",
       "  ENST00000361899       MT     8527-9207      + | ENST00000361899     MT-ATP6\n",
       "  ENST00000362079       MT     9207-9990      + | ENST00000362079      MT-CO3\n",
       "                          gene_id   gene_biotype     type\n",
       "                      <character>    <character> <factor>\n",
       "  ENSE00001489430 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00001536003 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00002160563 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00001750899 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00001489388 ENSG00000182378 protein_coding     exon\n",
       "              ...             ...            ...      ...\n",
       "  ENST00000361739 ENSG00000198712 protein_coding      cds\n",
       "  ENST00000361789 ENSG00000198727 protein_coding      cds\n",
       "  ENST00000361851 ENSG00000228253 protein_coding      cds\n",
       "  ENST00000361899 ENSG00000198899 protein_coding      cds\n",
       "  ENST00000362079 ENSG00000198938 protein_coding      cds\n",
       "  -------\n",
       "  seqinfo: 25 sequences (1 circular) from GRCh37 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Annotation(pbmc) <- annotations\n",
    "pbmc@assays$ATAC@annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06b7d2bf",
   "metadata": {},
   "source": [
    "## Compute Quality Control"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cc08827",
   "metadata": {},
   "source": [
    "Compute nucleosome signal score per cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c6f2092c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc <- NucleosomeSignal(pbmc)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54ecdd6b",
   "metadata": {},
   "source": [
    "Compute TSS enrichment score per cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0522ccb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Extracting TSS positions\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pbmc <- TSSEnrichment(object = pbmc, fast = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee229002",
   "metadata": {},
   "source": [
    "Add blacklist ratio and fraction of reads in peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ceb573e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc$blacklist_ratio <- pbmc$blacklist_region_fragments / pbmc$peak_region_fragments\n",
    "pbmc$pct_reads_in_peaks <- pbmc$peak_region_fragments / pbmc$passed_filters * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8aa5487",
   "metadata": {},
   "outputs": [],
   "source": [
    "View(pbmc@meta.data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12c385f0",
   "metadata": {},
   "source": [
    "## Visualizing Quality Control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e3a61c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(pbmc@meta.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9d0cc2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "a1 <- DensityScatter(pbmc, x = 'nCount_ATAC', y = 'TSS.enrichment', log_x = TRUE, quantiles = TRUE)\n",
    "a2 <- DensityScatter(pbmc, x = 'nucleosome_signal', y = 'TSS.enrichment', log_x = TRUE, quantiles = TRUE)\n",
    "\n",
    "a1 | a2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5347d759",
   "metadata": {},
   "outputs": [],
   "source": [
    "VlnPlot(object = pbmc, \n",
    "        features = c('nCount_ATAC', 'nFeature_ATAC', 'TSS.enrichment', 'nucleosome_signal', 'blacklist_ratio', 'pct_reads_in_peaks'),\n",
    "        pt.size = 0.1,\n",
    "        ncol = 6)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6786695",
   "metadata": {},
   "source": [
    "## Filtering poor quality cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b66cf14",
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc <- subset(x = pbmc,\n",
    "               subset = nCount_ATAC > 3000 &\n",
    "                 nCount_ATAC < 30000 &\n",
    "                 pct_reads_in_peaks > 15 & \n",
    "                 blacklist_ratio < 0.05 &\n",
    "                 nucleosome_signal < 4 &\n",
    "                 TSS.enrichment > 3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64be5984",
   "metadata": {},
   "source": [
    "## Normalization and linear dimensional reduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c054ed4",
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc <- RunTFIDF(pbmc) # normalization\n",
    "pbmc <- FindTopFeatures(pbmc, min.cutoff = 'q0') # selecting top features\n",
    "pbmc <- RunSVD(pbmc) # dimensionality reduction\n",
    "\n",
    "DepthCor(pbmc)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f53f7667",
   "metadata": {},
   "source": [
    "## Non-linear dimensional reduction and Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a92f8689",
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc <- RunUMAP(object = pbmc, reduction = 'lsi', dims = 2:30)\n",
    "pbmc <- FindNeighbors(object = pbmc, reduction = 'lsi', dims = 2:30)\n",
    "pbmc <- FindClusters(object = pbmc, algorithm = 3)\n",
    "\n",
    "DimPlot(object = pbmc, label = TRUE) + NoLegend()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
